---
title: "Batch correction across tissues and patients"
author: "Karol Nowicki-Osuch"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: paper
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Rscript -e "library(rmarkdown); render('Rebuttal_All.Rmd')"
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
# library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(umap)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(pheatmap)
# library(Seurat)
source("~/Dropbox/Postdoc/git/Gastric_IM//Analysis/Functions/auxiliary.R")
library(foreach)
library(doParallel)
```

# Combine data

Here I will combine processe data from our and Li studies. I will also introduce some metadata for sample tracking. 

```{r combine, eval=TRUE}
sce.selected <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_Cells_withBarcode_data.rds")

sce.list.filtered.tumor <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_filtered_Kumar_tumor_manual.rds")
sce.list.filtered.normal <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_filtered_Kumar_normal_manual.rds")

common.genes<-Reduce(intersect, list(rownames(sce.list.filtered.normal[[1]]),
                                     rownames(sce.list.filtered.tumor[[1]]),
                                     rowData(sce.selected)$Symbol)
)

gene.dictionary <-  rowData(sce.selected)[,1:2]
gene.dictionary <- gene.dictionary[gene.dictionary$Symbol %in% common.genes,]
gene.dictionary <- gene.dictionary[!duplicated(gene.dictionary$Symbol),]
# Fix row table for Li data
sce.list.filtered.normal <- lapply(sce.list.filtered.normal, function(n){
  # n <- sce.list.filtered.normal[[1]]
  n <- n[gene.dictionary$Symbol,] # Keep only genes that are common
  rowData(n) <- gene.dictionary
  # rowData(n)$Symbol <- rownames(n)
  rownames(n) <- rowData(n)$ID
  # rowData(n) <- rowData(n)[,c("ID", "Symbol")] # Keep only two columns about gene name
  
  n
})

sce.list.filtered.tumor <- lapply(sce.list.filtered.tumor, function(n){
  # n <- sce.list.filtered.normal[[1]]
  n <- n[gene.dictionary$Symbol,] # Keep only genes that are common
  rowData(n) <- gene.dictionary
  # rowData(n)$Symbol <- rownames(n)
  rownames(n) <- rowData(n)$ID
  # rowData(n) <- rowData(n)[,c("ID", "Symbol")] # Keep only two columns about gene name
  
  n
})

rowData(sce.selected) <- rowData(sce.selected)[,c("ID", "Symbol")]

# sce.selected <- sce.selected[rowData(sce.selected)$Symbol %in% common.genes,]
# match(common.genes, rowData(sce.selected)$Symbol)
# sce.selected <- sce.selected[rowData(sce.selected)$ID %in% rowData(sce.list.filtered.normal[[1]])$ID,]
sce.selected <- sce.selected[gene.dictionary$ID,]



# sce.selected <- sce.selected[match(common.genes, rowData(sce.selected)$Symbol),]


#######################
######################
#####################
####################
colData(sce.selected)$Cancer_type <- NA
colData(sce.selected)$Stage <- NA
colData(sce.selected)$Location <- NA
colData(sce.selected)$IM.Status <- NA
colData(sce.selected)$MMR_Status <- NA

# Remove total from the Kumar data
for (n in names(sce.list.filtered.normal)) {
  colData(sce.list.filtered.normal[[n]]) <- colData(sce.list.filtered.normal[[n]])[,colnames(colData(sce.list.filtered.normal[[n]])) != "total"]
  colnames(sce.list.filtered.normal[[n]]) <- NULL
}


# Remove total from the Kumar data
for (n in names(sce.list.filtered.tumor)) {
  colData(sce.list.filtered.tumor[[n]]) <- colData(sce.list.filtered.tumor[[n]])[,colnames(colData(sce.list.filtered.tumor[[n]])) != "total"]
  colnames(sce.list.filtered.tumor[[n]]) <- NULL
}

# Remove metadata from all data
metadata(sce.selected) <- list()
reducedDims(sce.selected) <- list()
assays(sce.selected) <- assays(sce.selected)[1]

```


```{r}
sce.list <- c(list(sample.our = sce.selected[,sce.selected$Patient == "Patient03"]), sce.list.filtered.normal)
sce <- do.call("cbind", sce.list)
sce <- do.call("cbind", sce.list.filtered.normal)
table(sce$Sample)
sce <- do.call("cbind", sce.list.filtered.tumor)
table(sce$Sample)

# Fix coldata that are factors into characters
i <- sapply(colData(sce), is.factor)
colData(sce)[i] <- lapply(colData(sce)[i], as.character)


# sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_with_Cancer.rds")
colData(sce) <- colData(sce)[,!colnames(colData(sce)) %in%  c("sizeFactor", "sum", "total", "detected")]
sce <- addPerFeatureQC(sce)
sce <- addPerCellQC(sce)
# colData(sce)[,colnames(colData(sce)) != "sizeFactor"]
# Batch normalization
sce <- batchelor::multiBatchNorm(sce, batch = sce[["Sample"]])



batch.order.names<-list()
length(batch.order.names) <- length(unique(paste(sce$Tissue, sce$Study, sep = "_")))
names(batch.order.names) <- c("NE_Nowicki-Osuch et al. 2022", "NG_Nowicki-Osuch et al. 2022", "Normal_Adjacent_Kumar at al. 2022",  "BSCJ_Nowicki-Osuch et al. 2022", "BE_Nowicki-Osuch et al. 2022")
# names(batch.order.names) <- c("NE_Nowicki-Osuch et al. 2022", "NSCJ_Nowicki-Osuch et al. 2022", "NG_Nowicki-Osuch et al. 2022", "NGB_Sathe et al. 2020", "Normal_Adjacent_Kumar at al. 2022", "ND_Nowicki-Osuch et al. 2022", "SMG_Nowicki-Osuch et al. 2022",  "BSCJ_Nowicki-Osuch et al. 2022", "GM_Nowicki-Osuch et al. 2022", "BE_Nowicki-Osuch et al. 2022","CIM_Nowicki-Osuch et al. 2022", "GIM_Nowicki-Osuch et al. 2022", "NAG_Zhang et al. 2019", "CAG_Zhang et al. 2019", "GIM_Zhang et al. 2019", "Ileum_Wang et. al. 2020", "Colon_Wang et. al. 2020", "Rectum_Wang et. al. 2020", "Gastric_Tumor_Kumar at al. 2022")

for (n in names(batch.order.names)) {
  batch.order.names[[n]]<-c(names(sort(table(sce$Sample[paste(sce$Tissue, sce$Study, sep = "_") == n]), decreasing = TRUE)))
}

corrected <- batch.correction.single(sce, batches = "Sample", m.order = batch.order.names, number.HVG = 2000)

# Save batch corrected counts in metdata
metadata(sce)$corrected_2000 <- corrected
# Save the object
# saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")
# Compute tsne on corrected counts
set.seed(50014)
# tsne <- Rtsne(t(corrected), pca = FALSE)

# cl <- makeCluster(7) #not to overload your computer
# registerDoParallel(cl)
# 
# corrected<-metadata(sce)$corrected_2000 
# results <-  foreach(i=c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)) %dopar% {
#   tmp.umap = umap::umap(t(corrected), min_dist = i, n_neighbors = 15) #calling a function
#   # tmp.umap <- uwot::umap(t(corrected), min_dist = i, verbose = F, n_threads = 24)
#   #do other things if you want
#   return(tmp.umap) 
# }
# stopCluster(cl)
# 
results.alt <-  foreach(i=c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)) %do% {
  # tmp.umap <- umap::umap(t(corrected), min_dist = i, n_neighbors = 15) #calling a function
  # i <- 0.2
  tmp.umap <- uwot::umap(t(corrected), min_dist = i, verbose = F, n_threads = 24, n_sgd_threads = "auto", batch = TRUE)
  #do other things if you want
  return(tmp.umap) 
}

names(results.alt)<-paste("umap_MinDist", c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), "N_Neighbors", 15, sep= "_" )
# reducedDims(sce)<-lapply(results, function (x) {x$layout[,1:2]})
reducedDims(sce) <- results.alt


# Visualise tissues
print(ggplot(data.frame(UMAP1 = reducedDims(sce)[[3]][,1],
                        UMAP2 = reducedDims(sce)[[3]][,2],
                        tissue = as.factor(colData(sce)$Tissue))) + 
        geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
        theme_cowplot(12) +
        # xlim(c(-20, 20))+
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
        scale_color_manual(values = colour.palet))
print(ggplot(data.frame(UMAP1 = reducedDims(sce)[[3]][,1],
                        UMAP2 = reducedDims(sce)[[3]][,2],
                        tissue = as.factor(sce$Celltypes_global))) + 
        geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
        theme_cowplot(12) #+
        # xlim(c(-20, 20))+
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
        # scale_color_manual(values = colour.palet))
)

```











```{r}
# Create list of all data
sce.list <- c(list(sample.our = sce.selected), sce.list.filtered.normal, sce.list.filtered.tumor)

# sce.list <- c(list(sample.our = sce.selected[,sce.selected$Patient == "Patient03"]), sce.list.filtered.normal[1])
# sce.list <- c(list(sample.our = sce.selected[,1:50]) ,sce.list.filtered.normal[1])
# Combine sce objects
sce <- do.call("cbind", sce.list)

# Fix coldata that are factors into characters
i <- sapply(colData(sce), is.factor)
colData(sce)[i] <- lapply(colData(sce)[i], as.character)

saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_with_Cancer.rds")


```

# Read and remove low quality samples

To be removed: 
1 Cancer sample from Li study
1 NG sample from Patient12 - it has unknown phenotype
1 NG sample from Patient05 - it has unknown phenotype

```{r fix-data, eval=TRUE}

# sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_with_Cancer.rds")
colData(sce) <- colData(sce)[,!colnames(colData(sce)) %in%  c("sizeFactor", "sum", "total", "detected")]
sce <- addPerFeatureQC(sce)
sce <- addPerCellQC(sce)
# colData(sce)[,colnames(colData(sce)) != "sizeFactor"]
# Batch normalization
sce <- batchelor::multiBatchNorm(sce, batch = sce[["Sample"]])

# ################## 
# # Introduce cell types from earleir analysis
# sce.all<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/All_sce_selected_corrected_withWang.rds")
# 
# sce$Celltypes20210625<-"NA"
# sce$Tissuetypes20210625<-"NA"
# matches <- match(paste(sce.all$Sample, sce.all$Barcode) ,paste(sce$Sample, sce$Barcode))
# matches<-matches[!is.na(matches)]
# matches2<-match(paste(sce$Sample, sce$Barcode) ,paste(sce.all$Sample, sce.all$Barcode))
# matches2<-matches2[!is.na(matches2)]
# sce$Celltypes20210625[matches] <- sce.all$Celltypes20210625[matches2]
# sce$Tissuetypes20210625[matches] <- sce.all$Tissuetypes20210625[matches2]

# Save the object
# saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected.rds")

```


# Perform batch correction

Here, I will perform batch correction across all samples. I treat each sample as an independent batch. They are sorted by the size of each sample.


```{r batch-correction-umap}
##############################################
# sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected.rds")

# Batch normalization
# sce<-batchelor::multiBatchNorm(sce, batch = sce[["Sample"]])


batch.order.names<-list()
length(batch.order.names) <- length(unique(paste(sce$Tissue, sce$Study, sep = "_")))
# names(batch.order.names) <- c("NE_Nowicki-Osuch et al. 2022", "NG_Nowicki-Osuch et al. 2022", "Normal_Adjacent_Kumar at al. 2022",  "BSCJ_Nowicki-Osuch et al. 2022", "BE_Nowicki-Osuch et al. 2022")
names(batch.order.names) <- c("NE_Nowicki-Osuch et al. 2022", "NSCJ_Nowicki-Osuch et al. 2022", "NG_Nowicki-Osuch et al. 2022", "NGB_Sathe et al. 2020", "Normal_Adjacent_Kumar at al. 2022", "ND_Nowicki-Osuch et al. 2022", "SMG_Nowicki-Osuch et al. 2022",  "BSCJ_Nowicki-Osuch et al. 2022", "GM_Nowicki-Osuch et al. 2022", "BE_Nowicki-Osuch et al. 2022","CIM_Nowicki-Osuch et al. 2022", "GIM_Nowicki-Osuch et al. 2022", "NAG_Zhang et al. 2019", "CAG_Zhang et al. 2019", "GIM_Zhang et al. 2019", "Ileum_Wang et. al. 2020", "Colon_Wang et. al. 2020", "Rectum_Wang et. al. 2020", "Gastric_Tumor_Kumar at al. 2022")

for (n in names(batch.order.names)) {
  batch.order.names[[n]]<-c(names(sort(table(sce$Sample[paste(sce$Tissue, sce$Study, sep = "_") == n]), decreasing = TRUE)))
}

corrected <- batch.correction.single(sce, batches = "Sample", m.order = batch.order.names, number.HVG = 2000)

# Save batch corrected counts in metdata
metadata(sce)$corrected_2000 <- corrected
# Save the object
saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")
# Compute tsne on corrected counts
set.seed(50014)
# tsne <- Rtsne(t(corrected), pca = FALSE)

# cl <- makeCluster(7) #not to overload your computer
# registerDoParallel(cl)
# 
# corrected<-metadata(sce)$corrected_2000 
# results <-  foreach(i=c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)) %dopar% {
#   tmp.umap = umap::umap(t(corrected), min_dist = i, n_neighbors = 15) #calling a function
#   # tmp.umap <- uwot::umap(t(corrected), min_dist = i, verbose = F, n_threads = 24)
#   #do other things if you want
#   return(tmp.umap) 
# }
# stopCluster(cl)
# 
results.alt <-  foreach(i=c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)) %do% {
  # tmp.umap <- umap::umap(t(corrected), min_dist = i, n_neighbors = 15) #calling a function
  # i <- 0.2
  tmp.umap <- uwot::umap(t(corrected), min_dist = i, verbose = F, n_threads = 24, n_sgd_threads = "auto", batch = TRUE)
  #do other things if you want
  return(tmp.umap) 
}

names(results.alt)<-paste("umap_MinDist", c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), "N_Neighbors", 15, sep= "_" )
# reducedDims(sce)<-lapply(results, function (x) {x$layout[,1:2]})
reducedDims(sce) <- results.alt
saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")

# Clustering on corrected data
g <- buildSNNGraph(corrected, k = 10)
clusters <- igraph::cluster_louvain(g)$membership


colour.palet <- c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))[1:length(unique(sce$Tissue))]
names(colour.palet) <- sort(unique(sce$Tissue))

# Visualise tissues
print(ggplot(data.frame(UMAP1 = reducedDims(sce)[[3]][,1],
                        UMAP2 = reducedDims(sce)[[3]][,2],
                        tissue = as.factor(colData(sce)$Tissue))) + 
        geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
        theme_cowplot(12) +
        # xlim(c(-20, 20))+
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
        scale_color_manual(values = colour.palet))
print(ggplot(data.frame(UMAP1 = reducedDims(sce)[[3]][,1],
                        UMAP2 = reducedDims(sce)[[3]][,2],
                        tissue = as.factor(sce$Celltypes_global))) + 
        geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
        theme_cowplot(12) #+
        # xlim(c(-20, 20))+
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
        # scale_color_manual(values = colour.palet))
)

print(ggplot(data.frame(UMAP1 = results[[2]]$layout[,1],
                        UMAP2 = results[[2]]$layout[,2],
                        tissue = as.factor(colData(sce)$Study))) + 
        geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
        theme_cowplot(12)# +
      # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
      # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout))))# +
      # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
      # scale_color_manual(values = colour.palet))
)

# Visualise cluster
print(ggplot(data.frame(UMAP1 = results[[3]]$layout[,1],
                        UMAP2 = results[[3]]$layout[,2],
                        clusters = as.factor(clusters))) + 
        theme_cowplot(12) +
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        geom_point(aes(UMAP1, UMAP2, colour = clusters), size = 0.5)
)

# Visualise cluster
print(ggplot(data.frame(UMAP1 = results[[2]]$layout[,1],
                        UMAP2 = results[[2]]$layout[,2],
                        clusters = as.factor(sce$Celltypes20210625))) + 
        theme_cowplot(12) +
        # ylim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        # xlim(c(floor(min(umap$layout)), ceiling(max(umap$layout)))) +
        geom_point(aes(UMAP1, UMAP2, colour = clusters), size = 0.5)
)



# Save clustering in new slot
colData(sce)$Global_cluster <- clusters
# saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/All_sce_corrected_selected.rds")


# # Perform differential expression
# markers <- findMarkers(sce, groups = colData(sce)$Global_cluster, 
#                        block = paste(colData(sce)$Patient, colData(sce)$Tissue, sep = "_"))
# 
# 
# markers.spec <- lapply(markers, function(n){
#   if(!is.na(n$Top[1]) & !is.nan(sum(as.matrix(n[1,4:ncol(n)])))){
#     test_n <- !is.na(n[1,4:ncol(n)])[1,]
#     cur_n <- n[n$FDR < 0.1 & apply(n[,4:ncol(n)], 1, function(x){sum(x[test_n] > 0)}) == sum(test_n),]
#     if(nrow(cur_n) > 0){
#       cur_n$GeneName <- rowData(sce)$Symbol[match(rownames(cur_n), rowData(sce)$ID)]
#     }
#   }
#   else{
#     cur_n <- NULL
#   }
#   cur_n
# })

# Save clustering results
# write.xlsx(markers.spec, paste("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Marker_genes/Global_filtered/Final/Marker_genes_combined.xlsx", sep = ""))

# Save the object
saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")
```



## Prepare data for visualisation

```{r}
library(anndata)
library(zellkonverter)
# sce.visualize<-sce
sce.visualize <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")
rowData(sce.visualize) <- rowData(sce.visualize)[,1:2]
sce.visualize$Global_cluster <- as.factor(sce.visualize$Global_cluster )
#randomise data
set.seed(50014)
jumbled<-sample(ncol(sce.visualize), replace = FALSE)
sce.visualize<-sce.visualize[,jumbled]


adata <- zellkonverter::SCE2AnnData(sce.visualize, X_name = "logcounts")
# adata$raw <- anndata::AnnData(counts(sce.visualize))
names(adata$obsm)[4] <- paste0(".",names(adata$obsm)[4])
names(adata$obsm)  <- paste0("X_",names(adata$obsm))
# adata$obsm <- adata$obsm[c(4,1:3,5:7)]

write_h5ad(adata, "/home/karolno/Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_Cells_with_Cancer.h5ad", compression = "gzip")

saveRDS(sce.visualize, "/home/karolno//Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_Cells_with_Cancer.rds")





```

## Manually transfer labels

```{r}
sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected.rds")
# I will only transfer tissue type labels. All I want them to be used for the analysis of samples using music
table(sce$Celltypes_global, sce$Global_cluster)
counts.per.cluster<-reshape2::dcast(as.data.frame(table(sce$Tissuetypes_global, sce$Global_cluster)), Var1 ~ Var2)
Tissue.types.by.cluster <- as.character(counts.per.cluster$Var1[unlist(apply(counts.per.cluster, 2, which.max))])

Tissue.types.by.cluster[9] <- "Cancer"
sce$Tissuetypes_global[sce$Study == "Kumar at al. 2022"] <- Tissue.types.by.cluster[sce$Global_cluster[sce$Study == "Kumar at al. 2022"]]



```

# SingleR label transfer

```{r}
library(SingleR)
library(tibble)
# run Training module of SingleR. Keep around 20 genes per cell type
SingleR.trained <- trainSingleR(ref=sce[,sce$Study != "Kumar at al. 2022"], labels=sce$Celltypes_global[sce$Study != "Kumar at al. 2022"], de.method="wilcox", de.n = 20)


library(BiocParallel)
param <- MulticoreParam(workers = 28) # run with all cores

# Run SingleR with single cell option and and use wilcox method for DE
prd.SingleR <- classifySingleR(test=sce[,sce$Study == "Kumar at al. 2022"] , trained=SingleR.trained, BPPARAM = param)
# prd.SingleR <- prd.SingleR.BE
table(prd.SingleR$labels)
sce$Celltypes_global[sce$Study == "Kumar at al. 2022"] <- prd.SingleR$labels
sce$Celltypes_global[sce$Global_cluster == 9] <- "Cancer"
```


```{r}
saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected_labels.rds")
sce.visualize <- sce
rowData(sce.visualize) <- rowData(sce.visualize)[,1:2]
sce.visualize$Global_cluster <- as.factor(sce.visualize$Global_cluster )
sce.visualize$Celltypes_global <- as.factor(sce.visualize$Celltypes_global)
#randomise data
set.seed(50014)
jumbled<-sample(ncol(sce.visualize), replace = FALSE)
sce.visualize<-sce.visualize[,jumbled]


adata <- zellkonverter::SCE2AnnData(sce.visualize, X_name = "logcounts")
# adata$raw <- anndata::AnnData(counts(sce.visualize))
names(adata$obsm)[4] <- paste0(".",names(adata$obsm)[4])
names(adata$obsm)  <- paste0("X_",names(adata$obsm))
# adata$obsm <- adata$obsm[c(4,1:3,5:7)]

write_h5ad(adata, "/home/karolno/Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_Cells_with_Cancer_with_Cells.h5ad", compression = "gzip")

saveRDS(sce.visualize, "/home/karolno//Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_Cells_with_Cancer_with_Cells.rds")




```


# Get Figures

```{r}
sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected_labels.rds")

for.printing <- data.frame(UMAP1 = reducedDims(sce)$umap_MinDist_0.2_N_Neighbors_15 [,1],
                           UMAP2 = reducedDims(sce)$umap_MinDist_0.2_N_Neighbors_15 [,2],
                           Tissue = factor(sce$Tissue, levels = names(colour.palet)),
                           # Patient = as.factor(sce$Patient),
                           Study = as.factor(sce$Study),
                           # Time = as.factor(sce$Time),
                           # CellType = factor(sce$Celltypes_global, levels = names(cell.order)),
                           Tissuetype = as.factor(sce$Tissuetypes_global)
                           # Patientstatus = as.factor(sce$Patient_status),
                           # Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           # Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           # Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)
#randomise data
set.seed(50014)
jumbled<-sample(row(for.printing), replace = FALSE)
for.printing<-for.printing[jumbled,]



All.cells.umap <- ggplot(data = for.printing) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  # scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
  ggtitle("All Cells with Cancer")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) 
  # ylim(umap.min.max) +
  # xlim(umap.min.max)
print(All.cells.umap)


ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/All.umap.tiff", All.cells.umap, device = "tiff", units="in", width=12, height=10, bg = "white")
Tissue.type.colours <-c(
  "Squamous"  = "#990F0F",
  "Immune" = "#422CB2",
  "Connective" = "#0F6B99",
  "Columnar" = "#6B990F",
  "Other" = "grey60",
  "Glandular" = "#FFB2B2",
  "Cancer" = "#99540F"
)

All.cells.umap.tissuetype <- ggplot(data = for.printing) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissuetype), size = 0.5) +
  scale_color_manual(values = Tissue.type.colours) + 
  # scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
  ggtitle("All Cells with Cancer Cell type")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) 
  # ylim(umap.min.max) +
  # xlim(umap.min.max)
print(All.cells.umap.tissuetype)


ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/All.umap.tissuetype.tiff", All.cells.umap.tissuetype, device = "tiff", units="in", width=12, height=10, bg = "white")


i <- c("Kumar at al. 2022")


for.printing.tmp<-for.printing[c(which(!(for.printing$Study %in% i)),which(for.printing$Study %in% i)), ]
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Study %in% i, "Cancer Data", "other")


Cancer.tissues <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(!(for.printing.tmp$Study %in% i))), rep(1, times = sum(for.printing.tmp$Study %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
  ggtitle("Cancer Data")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4)))
  # ylim(umap.min.max) +
  # xlim(umap.min.max)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Cancer_samples.umap.tiff", Cancer.tissues, device = "tiff", units="in", width=12, height=10, bg = "white")


```


# Music analysis 


```{r}
## Run MuSiC, no intermediate with SMG and NE
library(MuSiC)

sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected_labels.rds")
sce.normal<-sce[,sce$Tissuetypes_global %in% c("Columnar", "Glandular", "Squamous") & sce$Study != "Kumar at al. 2022"]
sce.normal <- sce.normal[, (sce.normal$Tissue %in% c("ND", "Ileum", "Colon") & sce.normal$Celltypes_global %in% c("Enterocytes", "Enterocytes_Intermediate", "Intestinal_Undifferentiated", "Goblet", "Enteroendocrine_CHGA" )) |
                    (sce.normal$Tissue %in% c("NGB", "NG") & sce.normal$Celltypes_global %in% c("Foveolar_Differentiated", "Foveolar_Intermediate", "Neck-Cells", "Enteroendocrine_CHGA", "Enteroendocrine_GHRL", "Parietal", "Chief" )) |
                    (sce.normal$Tissue %in% c("SMG") & sce.normal$Celltypes_global %in% c("SMG" )) |
                    (sce.normal$Tissue %in% c("NE") & sce.normal$Celltypes_global %in% c("Basal", "Suprabasal", "Intermediate", "Superficial")) |
                    (sce.normal$Tissue %in% c("NSCJ") & sce.normal$Celltypes_global %in% c("SMG" ))
]



sce.normal$Celltypes_global[sce.normal$Celltypes_global %in% c("Foveolar_Differentiated", "Foveolar_Intermediate")]<-"Foveolar"
sce.normal$Celltypes_global[sce.normal$Celltypes_global %in% c("Enterocytes_Intermediate", "Enterocytes")]<-"Enterocytes"

#create phenotype matrix
pheno.matrix<-colData(sce.normal)[,c("Sample", "Patient", "Tissue", "Celltypes_global")]
pheno.matrix$Celltype.tissue <- pheno.matrix$Tissue

pheno.matrix$Celltype.tissue[pheno.matrix$Tissue %in%  c("ND", "Ileum")] <- "Intestine"
pheno.matrix$Celltype.tissue[pheno.matrix$Tissue %in%  c("NG", "NGB")] <- "Gastric"
# pheno.matrix$Celltype.tissue[pheno.matrix$Tissue == "Colon"] <- "Colon"
pheno.matrix$Celltype.tissue<-paste(pheno.matrix$Celltype.tissue, pheno.matrix$Celltypes_global, sep = "_")
#annotate phenotypes
metadata <- data.frame(labelDescription= c("Sample", "Patient", "Tissue", "Cell Type", "Cell Type Combined"), row.names=colnames(pheno.matrix))

#create Expression set for the input into MuSiC
cur.sce1<-ExpressionSet(assayData = as.matrix(counts(sce.normal)), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix), varMetadata = metadata) )

# sce to analyse
sce.cancer <- sce[,sce$Tissuetypes_global %in% c("Columnar", "Cancer") & sce$Study == "Kumar at al. 2022"]
# 
# Get data for the test samples, all data is contained in sce
# sce.BE.test.MuSiC<-sce
# sce.BE.test.MuSiC<-sce[,sce$Celltypes_global %in% c("Enterocytes", "Enterocytes_Intermediate", "Foveolar_Differentiated", "Foveolar_Intermediate", "Intestinal_Undifferentiated", "Neck-Cells" )]
# jumbled_MuSiC <- sample(ncol(sce.BE.test.MuSiC))
# jumbled_MuSiC <- readRDS("~/Dropbox/Postdoc/2019-10-29_Gastric_IM/Results/MuSiC_randomno.RDS")

n.threat<-24 #number of CPU
ncells<-ceiling(seq(0, ncol(sce.cancer), length.out = n.threat+1))  # number of cells per CPU to be analysed
# Around 4.5K cells
# ncells <- 5000

# sce.BE.test.MuSiC<-sce.BE.test.MuSiC[,jumbled_MuSiC[1:ncells]]
pheno.matrix.test <-colData(sce.cancer)[,c("Sample", "Patient", "Tissue", "Celltypes_global")]

metadata2<-data.frame(labelDescription= c("Sample", "Patient", "Tissue", "Celltypes_global"), row.names=colnames(pheno.matrix.test))


```

```{r, eval = FALSE}
# Run MuSiC in parallel
# Compute tsne on corrected counts
set.seed(50014)

# tsne <- Rtsne(t(corrected), pca = FALSE)

# Split single cell object into small files and save them as individual .Rdata files. See option 3 here: https://lcolladotor.github.io/2013/11/14/reducing-memory-overhead-when-using-mclapply/#.YPbxgOy8g-U
# it makes the memory footprint smallest in principle
for (i in c(1:n.threat)) {
  data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce.cancer[,(ncells[i]+1):(ncells[i+1])])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):(ncells[i+1]),]), varMetadata = metadata2))
  save(data.all.test, file = paste0("chunk", i, ".Rdata"))
}

rm(sce)
rm(sce.cancer)
rm(sce.normal)
rm(tpm.data)
rm(sce.visualize)
rm(adata)
rm(SingleR.trained)

gc()

run.music<-function(x) {
  # print(i)
  load(paste0("chunk", x, ".Rdata"))
  cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltype.tissue", samples = "Sample", select.ct = unique(cur.sce1$Celltype.tissue), verbose = FALSE)
  return(cell.prop)
}


results.no.intermediate<-mclapply(1:n.threat, 
                                  FUN = run.music, 
                                  mc.cores = 2)
for (i in c(1:n.threat)) {
  fn<-paste0("chunk", i, ".Rdata")
  if (file.exists(fn)) {
    #Delete file if it exists
    file.remove(fn)
  }
}


# cl <- makeCluster(2) #not to overload your computer
# registerDoParallel(cl)

# results.no.intermediate <-  foreach(i = c(1:n.threat)) %dopar% {
# results.no.intermediate <-  foreach(i = c(1:n.threat)) %do% {
#   library(MuSiC)
#   library(scran)
#   #create expression set
#   # data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce.BE.test.MuSiC[,(ncells[i]+1):10])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):10,]), varMetadata = metadata2))
#   data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce[,(ncells[i]+1):(ncells[i+1])])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):(ncells[i+1]),]), varMetadata = metadata2))
#   cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltype.tissue", samples = "Sample", select.ct = unique(cur.sce1$Celltype.tissue), verbose = FALSE)
#   #do other things if you want
#   return(cell.prop) 
# }
# stopCluster(cl)
estimated.weights.no.intermediate<-do.call(rbind,lapply(results.no.intermediate, function (x) return(x$Est.prop.weighted)))
# cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltypes_global", samples = "Patient", select.ct = unique(cur.sce1$cell.types), verbose = FALSE)
# cell.prop.old<-cell.prop
# cell.prop<-cell.prop.all

saveRDS(results.no.intermediate, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/MuSiC_Cancer.rds")
# saveRDS(jumbled_MuSiC, "~/Dropbox/Postdoc/2019-10-29_Gastric_IM/Results/MuSiC_randomno.rds")
```



```{r}
sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_with_Cancer_corrected_labels.rds")

sce.cancer <- sce[,sce$Tissuetypes_global %in% c("Columnar", "Cancer") & sce$Study == "Kumar at al. 2022"]

colour.palet <- c("NE" = "#990F0F", 
                  "NSCJ" = "#CC5151",
                  "SMG" = "#FFB2B2",
                  "GM" = "#E5B17E",
                  "BE" = "#99540F",
                  "BSCJ" = "#CC8E51",
                  "NG" = "#6B990F",
                  "NGB" = "#A3CC51",
                  "Normal_Adjacent" = "#A3CC51",
                  "Gastric_Tumor"  = "darkred",
                  "NAG" = "#0F6B99",
                  "CAG" = "#2C85B2",
                  "GIM" = "#51A3CC",
                  "CIM" = "#7EC3E5",
                  "ND" = "#260F99",
                  "Ileum" = "#422CB2",
                  "Colon" = "#6551CC",
                  "Rectum" = "#8F7EE5")
# colour.palet <- c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))[1:length(unique(sce$Tissue))]
# names(colour.palet) <- sort(unique(sce$Tissue))
cell.order<-c(colorBlindness::SteppedSequential5Steps[6:10],rev(RColorBrewer::brewer.pal(n = 8,name = "Reds")) ,colorBlindness::SteppedSequential5Steps[c(11:13,15, 16,18, 21:25)], "grey60")

names(cell.order)<-c(
  "Superficial", 
  "Intermediate",    
  "Suprabasal",
  "Basal",
  "SMG",
  
  "Foveolar_Differentiated",
  "Foveolar_Intermediate",
  "Neck-Cells",
  "Chief",        
  "Parietal",
  "Enteroendocrine_CHGA",
  "Enteroendocrine_GHRL", 
  "Enteroendocrine_GAST",
  
  
  
  "Enterocytes",
  "Enterocytes_Intermediate",
  "Intestinal_Undifferentiated",
  "Goblet",
  
  "Fibroblasts",
  "Endothelial",
  
  "B-Cells",
  "T-Cells",
  "Plasma-Cells",
  "Macrophages",
  "Mast",
  
  "Erythrocyte"
)

for.printing <- data.frame(
  Tissue = factor(sce.cancer$Tissue, levels = names(colour.palet)),
  Cluster = as.factor(sce.cancer$Global_cluster),
  Patient = as.factor(sce.cancer$Patient),
  CellType = factor(sce.cancer$Celltypes_global, levels = names(cell.order)),
  Squamous = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("NE", colnames(estimated.weights.no.intermediate))]), 
  SMG = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("SMG_SMG", colnames(estimated.weights.no.intermediate)), drop = FALSE]), 
  Transitional = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("NSCJ_SMG", colnames(estimated.weights.no.intermediate)), drop = FALSE]), 
  Colon = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Colon", colnames(estimated.weights.no.intermediate))]), 
  Gastric = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Gastric", colnames(estimated.weights.no.intermediate))]), 
  Intestine = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Intestine", colnames(estimated.weights.no.intermediate))])
  
)
for.printing.long<-reshape2::melt(for.printing, variable.name = "Contributor", value.name = "Contribution" )

# for.printing.long<-reshape2::melt(for.printing, id.vars = c(), variable.name = "Contributor", value.name = "Contribution" )
for.printing.long$Contributor <- factor(for.printing.long$Contributor, levels = c("Squamous", "SMG", "Transitional", "Gastric", "Intestine", "Colon"))
for.printing.long$Tissue <- factor(for.printing.long$Tissue, levels = names(colour.palet))
for.printing.long$Tissue_type <- NA
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NSCJ", "NG", "NGB", "Normal_Adjacent")] <- "Stomach"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("ND", "Ileum", "Colon", "Rectum")] <- "Intestine"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("BE", "BSCJ", "GM")] <- "Esophageal IM"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NAG", "CAG", "GIM", "CIM")] <- "Gastric IM"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("Gastric_Tumor")] <- "Gastric Cancer"
for.printing.long$Tissue_type <- factor(for.printing.long$Tissue_type, levels = c("Stomach", "Intestine", "Esophageal IM", "Gastric IM", "Gastric Cancer"))
for.printing.long$is.cancer.cell <- factor(ifelse(for.printing.long$Cluster == 9, "Cancer", "Normal"), levels = c("Normal", "Cancer"))
colour.palet.contributor <- c("Squamous" = "#990F0F", 
                              "Transitional" = "#CC5151",
                              "SMG" = "#FFB2B2",
                              "Gastric" = "#6B990F",
                              "Intestine" = "#260F99",
                              "Colon" = "#6551CC")
Violin.plot.all <- ggplot(data = for.printing.long, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x")
Violin.plot.all

Violin.plot.all.is.cancer <- ggplot(data = for.printing.long, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(is.cancer.cell), drop = TRUE,scales = "free_x")
Violin.plot.all.is.cancer
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Cancer_Contribution.pdf", Violin.plot.all.is.cancer, device = "pdf", units="in", width=6, height=4, useDingbats = FALSE)


Violin.plot.cancer <- ggplot(data = for.printing.long[for.printing.long$Cluster == 9,], aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x")
Violin.plot.cancer

Violin.plot.other <- ggplot(data = for.printing.long[for.printing.long$Cluster != 9,], aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x")
Violin.plot.other

Violin.plot.other.patient <- ggplot(data = for.printing.long[for.printing.long$Cluster != 9,], aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Patient), drop = TRUE,scales = "free_x")
Violin.plot.other.patient
Violin.plot.all.patient <- ggplot(data = for.printing.long, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Patient, is.cancer.cell), drop = TRUE,scales = "free_x")
Violin.plot.all.patient


```


#Monocle analysis

```{r}
library(monocle3)
# Load data 
# sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")
sce <- sce[,sce$Celltypes_global %in% c("Enterocytes_Intermediate", "Enterocytes", "Neck-Cells", "Foveolar_Differentiated", "Intestinal_Undifferentiated", "Foveolar_Intermediate", "Cancer")]

# Create monocle3 object
counts.data<-counts(sce)
colnames(counts.data)<-paste(colData(sce)$Barcode, colData(sce)$Sample, sep = "_")
cell.data<-colData(sce)
rownames(cell.data)<-paste(colData(sce)$Barcode, colData(sce)$Sample, sep = "_")
gene.data<-rowData(sce)
gene.data$gene_short_name<-gene.data$Symbol
cds <- new_cell_data_set(counts.data,
                         cell_metadata = cell.data,
                         gene_metadata = gene.data)

#Get PC
cds <- preprocess_cds(cds, num_dim = 50)

# Batch correct
cds <- align_cds(cds, alignment_group = "Sample", residual_model_formula_str = "~ MT.prop" )

# Get umap
cds <- reduce_dimension(cds)

# plot using building method
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "Celltypes_global")
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "Tissue")

# Visualise Clusters
CellTypes.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                  UMAP2 = reducedDims(cds)$UMAP[,2],
                                  CellTypes = factor(cds$Celltypes_global, levels = names(cell.order)[names(cell.order) %in% unique(colData(cds)$Celltypes_global)]),
                                  Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"), "Normal", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = CellTypes, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Normal" = 1 , "Other" = 0.00)) +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=2))) +
  scale_color_manual(values = cell.order[names(cell.order) %in% unique(colData(cds)$Celltypes_global)])
# +
#   geom_text(data= aggregate(reducedDims(cds)$UMAP, list(cds$Fibroblast_clusters), mean),
#             mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(CellTypes.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_CellTypes.tiff", CellTypes.plot, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_CellTypes.pdf", CellTypes.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

# Visualise Clusters
Tissue.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                               UMAP2 = reducedDims(cds)$UMAP[,2],
                               Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue), size = 0.5) + 
  theme_void() +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])

# +
#   geom_text(data= aggregate(reducedDims(cds)$UMAP, list(cds$Fibroblast_clusters), mean),
#             mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(Tissue.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue.tiff", Tissue.plot, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue.pdf", Tissue.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

Tissue.plot.normal<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]),
                                      Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"), "Normal", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Normal" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])


print(Tissue.plot.normal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_normal.tiff", Tissue.plot.normal, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_normal.pdf", Tissue.plot.normal, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

Tissue.plot.IM<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]),
                                      Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"), "Other", "IM/Cancer")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum", "Normal_Adjacent"),0.01,1)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("IM/Cancer" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])


print(Tissue.plot.IM)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_IM.tiff", Tissue.plot.IM, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_IM.pdf", Tissue.plot.IM, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)


Cancer.study.only<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]),
                                      Alpha = ifelse(cds$Study %in% c("Kumar at al. 2022"), "Cancer", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Study %in% c("Kumar at al. 2022"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Cancer" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])


print(Cancer.study.only)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_Cancer_only.tiff", Cancer.study.only, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_Cancer_only.pdf", Cancer.study.only, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)
factor(ifelse(for.printing.long$Cluster == 9, "Cancer", "Normal"), levels = c("Cancer-unlike", "Cancer-like"))

Cancer.study.only.is.cancer<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(ifelse(cds$Global_cluster == 9, "Cancer-like", "Cancer-unlike"), levels = c("Cancer-unlike", "Cancer-like")),
                                      Alpha = ifelse(cds$Study %in% c("Kumar at al. 2022"), "Cancer", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Study %in% c("Kumar at al. 2022"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Cancer" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = c("Cancer-like" = "darkred", "Cancer-unlike" = "green"))


print(Cancer.study.only.is.cancer)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_Cancer_only.is.cancer.tiff", Cancer.study.only.is.cancer, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_Tissue_Cancer_only.is.cancer.pdf", Cancer.study.only, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)




cds <- cluster_cells(cds)
plot_cells(cds, color_cells_by = "partition", group_label_size = 8)

# Visualise Partitions 
partitions.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                   UMAP2 = reducedDims(cds)$UMAP[,2],
                                   Partition = factor(partitions(cds)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Partition), size = 0.5) + 
  theme_void() +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=2, override.aes = list(size=4))) +
  geom_text(data= aggregate(reducedDims(cds)$UMAP, list(partitions(cds)), mean),
            mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(partitions.plot)




cds <- learn_graph(cds)

plot_cells(cds,
           color_cells_by = "Celltypes_global",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)

cds <- order_cells(cds)


pseudotime.plot <- plot_cells(cds,
                              color_cells_by = "pseudotime",
                              label_cell_groups=FALSE,
                              label_leaves=FALSE,
                              label_branch_points=FALSE,
                              graph_label_size=5,
                              cell_size = .75) + theme_void()
plot(pseudotime.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_pseudotime.tiff", pseudotime.plot, device = "tiff", units="in", width=8, height=6, bg = "white")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Monocle_umap_pseudotime.pdf", pseudotime.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

# colour.palet <- c("NE" = "#990F0F", 
#                   "NSCJ" = "#CC5151",
#                   "SMG" = "#FFB2B2",
#                   "NG" = "#6B990F",
#                   "NGB" = "#A3CC51",
#                   "GM" = "#E5B17E",
#                   "BE" = "#99540F",
#                   "BSCJ" = "#CC8E51",
#                   "NAG" = "#0F6B99",
#                   "CAG" = "#2C85B2",
#                   "GIM" = "#51A3CC",
#                   "CIM" = "#7EC3E5",
#                   "ND" = "#260F99",
#                   "Ileum" = "#422CB2",
#                   "Colon" = "#6551CC",
#                   "Rectum" = "#8F7EE5")
to.use <- colData(cds)$Tissuetypes_global %in% c("Columnar", "Cancer") &  !colData(cds)$Tissue %in% c("NE", "SMG") & !(colData(cds)$Global_cluster == 9 & !(colData(cds)$Tissue %in% c("Normal_Adjacent", "Gastric_Tumor")))
colData(cds)$Patient_status.alt  <- colData(cds)$Patient_status
colData(cds)$Patient_status.alt[colData(cds)$Global_cluster == 9] <- "Cancer-like"

Violin.plot.all <- ggplot(data = data.frame(Tissue = factor(colData(cds)$Tissue, levels = (names(colour.palet))),
                                            Tissue_type = factor(colData(cds)$Patient_status.alt, levels = c("Healthy", "BE", "IM","Gastric Cancer", "Cancer-like")),
                                            Pseudotime = pseudotime(cds))[to.use,], 
                          aes(x = Tissue, y = Pseudotime)) +
  theme_cowplot(12) +
  # ylim(0,1) +
  geom_violin(aes(fill = Tissue), scale = "area", trim = TRUE) + 
  scale_fill_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)]) +
  # scale_fill_manual(values = colour.palet.contributor) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  facet_grid( . ~ Tissue_type, drop = TRUE, scales = "free_x", space = "free")# + coord_flip()

plot(Violin.plot.all)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Cancer/Violin_pseudotime.pdf", Violin.plot.all, device = "pdf", units="in", width=15, height=8, useDingbats = FALSE)



saveRDS(cds, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/monocle_cancer_data.rds")

```



# End Matter

To finish get session info:

```{r Endnote, include=TRUE, echo=TRUE, warning=FALSE, eval=TRUE, message=FALSE, tidy=TRUE}
sessionInfo()
```

