---
title: "Comparison of columnar cell types"
author: "Karol Nowicki-Osuch"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: paper
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(colorBlindness)
library(ggplot2)
library(SingleR)
library(Rcpp)
library(RcppEigen)
library(ggsci)
library(scibet)
library(tidyverse)
# library(scater)
# library(DropletUtils)
library(openxlsx)
# library(Rtsne)
# library(umap)
library(RColorBrewer)
library(viridis)
library(cowplot)
# library(pheatmap)
# library(Seurat)
source("~/Dropbox/Postdoc/git/Gastric_IM//Analysis/Functions/auxiliary.R")
library(scater)
library(TSCAN)
# library(foreach)
library(doParallel)
library(MuSiC)

```

# Set up

## Read in data and get colour schemes

```{r visualise}
sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_after_exclusion2.rds")

# Keep only columnar cell types
corrected<-metadata(sce)$corrected_2000[,sce$Tissuetypes_global == "Columnar" & !(sce$Tissue %in% c("SMG", "NE"))]
sce <- sce[,sce$Tissuetypes_global == "Columnar" & !(sce$Tissue %in% c("SMG", "NE"))]

# Recompute umap

# Compute tsne on corrected counts
set.seed(50014)
# tsne <- Rtsne(t(corrected), pca = FALSE)

cl <- makeCluster(7) #not to overload your computer
registerDoParallel(cl)

results <-  foreach(i=c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)) %dopar% {
  tmp.umap = umap::umap(t(corrected), min_dist = i, n_neighbors = 15) #calling a function
  #do other things if you want
  return(tmp.umap) 
}
#foreach(j=c(15, 100)) %:%
stopCluster(cl)

names(results)<-paste("umap_MinDist", c(0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), "N_Neighbors", 15, sep= "_" )
reducedDims(sce)<-lapply(results, function (x) {x$layout[,1:2]})




#Taken from colorBlindness::SteppedSequential5Steps
## colorBlindness::displayAvailablePalette(color="white")
colour.palet <- c("NE" = "#990F0F", 
                  "NSCJ" = "#CC5151",
                  "SMG" = "#FFB2B2",
                  "GM" = "#E5B17E",
                  "BE" = "#99540F",
                  "BSCJ" = "#CC8E51",
                  "NG" = "#6B990F",
                  "NGB" = "#A3CC51",
                  "NAG" = "#0F6B99",
                  "CAG" = "#2C85B2",
                  "GIM" = "#51A3CC",
                  "CIM" = "#7EC3E5",
                  "ND" = "#260F99",
                  "Ileum" = "#422CB2",
                  "Colon" = "#6551CC",
                  "Rectum" = "#8F7EE5")
# colour.palet <- c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))[1:length(unique(sce$Tissue))]
# names(colour.palet) <- sort(unique(sce$Tissue))
cell.order<-c(colorBlindness::SteppedSequential5Steps[6:10],rev(RColorBrewer::brewer.pal(n = 8,name = "Reds")) ,colorBlindness::SteppedSequential5Steps[c(11:13,15, 16,18, 21:25)], "grey60")

names(cell.order)<-c(
  "Superficial", 
  "Intermediate",    
  "Suprabasal",
  "Basal",
  "SMG",
  
  "Foveolar_Differentiated",
  "Foveolar_Intermediate",
  "Neck-Cells",
  "Chief",        
  "Parietal",
  "Enteroendocrine_CHGA",
  "Enteroendocrine_GHRL", 
  "Enteroendocrine_GAST",
  
  
  
  "Enterocytes",
  "Enterocytes_Intermediate",
  "Intestinal_Undifferentiated",
  "Goblet",
  
  "Fibroblasts",
  "Endothelial",
  
  "B-Cells",
  "T-Cells",
  "Plasma-Cells",
  "Macrophages",
  "Mast",
  
  "Erythrocyte"
)
set.seed(12345)
jumbled<-sample(ncol(sce), replace = FALSE)
sce<-sce[,jumbled]

# Clustering on corrected data in the selected tissues
g <- buildSNNGraph(corrected, k = 10)
clusters_selected <- igraph::cluster_louvain(g)$membership
clusters_selected <- clusters_selected[jumbled]
clusters <- sce$Global_cluster_selected

# Perform differential expression
markers <- findMarkers(sce, groups = clusters_selected, 
                       block = paste(colData(sce)$Patient, colData(sce)$Tissue, sep = "_"))


markers.spec <- lapply(markers, function(n){
  if(!is.na(n$Top[1]) & !is.nan(sum(as.matrix(n[1,4:ncol(n)])))){
    test_n <- !is.na(n[1,4:ncol(n)])[1,]
    cur_n <- n[n$FDR < 0.1 & apply(n[,4:ncol(n)], 1, function(x){sum(x[test_n] > 0)}) == sum(test_n),]
    if(nrow(cur_n) > 0){
      cur_n$GeneName <- rowData(sce)$Symbol[match(rownames(cur_n), rowData(sce)$ID)]
    }
  }
  else{
    cur_n <- NULL
  }
  cur_n
})

# Save clustering results
write.xlsx(markers.spec, paste("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Marker_genes/Columnar/Marker_genes_combined_clean.xlsx", sep = ""))

sce$Columnar_clusters<-clusters_selected
# 
# reannot.clusters<-read.delim("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Additional_files/CellType_Global_clusters_20210702.txt")
# 
# sce$Celltypes_global<-reannot.clusters$CellType[match(sce$Global_cluster_selected, reannot.clusters$GlobalCluster)]


```

## Prepare data for visualisation

```{r}

sce.visualize<-sce
logcounts(sce.visualize)<-NULL
# randomise expression values nad moved htem to log2 space
counts(sce.visualize)<-logcounts(sce)#[,jumbled]

# Move Symbol to row name to keep it easy to loead into the data
rownames(sce.visualize)<-rowData(sce.visualize)$Symbol

colData(sce.visualize)<-colData(sce.visualize)[,c(4, 3, 1, 10:14, 16, 17,  19:27, 18, 8:9, 15)]
for(n in 1:(ncol(colData(sce.visualize))-4)) {
  colData(sce.visualize)[,n] <-  as.factor(colData(sce.visualize)[,n])
}   

for(n in (ncol(colData(sce.visualize))-3):(ncol(colData(sce.visualize))-1)) {
  colData(sce.visualize)[,n] <- log2(colData(sce.visualize)[,n] )
}




sceasy::convertFormat(sce.visualize, from="sce", to="anndata",
                      outFile="/home/karolno/Dropbox/Postdoc/2021-05-19_Gastric_IM/Visualisation/All_sce_selected_corrected_Columnar.h5ad")

saveRDS(sce, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")

```


## Plot figures

After cosntructing UMAP projections I will plot them to see what looks best and how the data behave. I will also take a look at the new clustering data.

```{r}
sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")

umap.data<-reducedDims(sce)$umap_MinDist_0.2_N_Neighbors_15 
umap.min.max<-c(floor(min(umap.data[,1:2])), ceiling(max(umap.data[,1:2])))

# umap.data<-reducedDims(sce)$fle
# umap.min.max<-c(floor(min(fle.data[,1:2])), ceiling(max(fle.data[,1:2])))

ggplot(data.frame(UMAP1 = umap.data[,1],
                  UMAP2 = umap.data[,2],
                  Tissue = factor(sce$Celltypes_global, levels = names(cell.order)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue), size = 0.5) + 
  theme_void() +
  ylim(umap.min.max) +
  xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
  scale_color_manual(values = cell.order) 

# Visualise cluster
print(ggplot(data.frame(UMAP1 = umap.data[,1],
                        UMAP2 = umap.data[,2],
                        clusters = as.factor(sce$Columnar_clusters))) + 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max) +
        geom_point(aes(UMAP1, UMAP2, colour = clusters), size = 0.5))

# Visualise cluster
print(ggplot(data.frame(UMAP1 = umap.data[,1],
                        UMAP2 = umap.data[,2],
                        clusters = as.factor(sce$Global_cluster_selected))) + 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max) +
        geom_point(aes(UMAP1, UMAP2, colour = clusters), size = 0.5))

# Visualise tissues
tissues.plot<-ggplot(data.frame(UMAP1 = umap.data[,1],
                                UMAP2 = umap.data[,2],
                                tissue = factor(colData(sce)$Tissue, levels = names(colour.palet)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
  theme_cowplot(12) +
  ylim(umap.min.max) +
  xlim(umap.min.max) +
  # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
  scale_color_manual(values = colour.palet)

tissues.plot<-ggplot(data.frame(UMAP1 = umap.data[,1],
                                UMAP2 = umap.data[,2],
                                tissue = factor(colData(sce)$Tissue, levels = names(colour.palet)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = tissue), size = 0.5) + 
  theme_void() +
  ylim(umap.min.max) +
  xlim(umap.min.max) +
  # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
  scale_color_manual(values = colour.palet)

print(tissues.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Allumap.tiff", tissues.plot, device = "tiff", units="in", width=11, height=10)

#Plot only BE cells

for.printing <- data.frame(UMAP1 = umap.data[,1],
                           UMAP2 = umap.data[,2],
                           Tissue = factor(sce$Tissue, levels = names(colour.palet)),
                           # Patient = as.factor(sce$Patient),
                           # Study = as.factor(sce$Study),
                           # Time = as.factor(sce$Time),
                           CellType = factor(sce$Celltypes_global, levels = names(cell.order)),
                           MUC5AC = logcounts(sce)[rowData(sce)$Symbol == "MUC5AC",],
                           FABP1 = logcounts(sce)[rowData(sce)$Symbol == "FABP1",],
                           GPA33 = logcounts(sce)[rowData(sce)$Symbol == "GPA33",]
                           # Tissuetype = as.factor(sce$Tissuetypes_global),
                           # Patientstatus = as.factor(sce$Patient_status),
                           # Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           # Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           # Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)

i <- c("GM", "BE", "BSCJ")
ii <- "Esophagus"
for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Esophageal tissues

BE.tissues <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("BE")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Eso_tissue.tiff", BE.tissues, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$MUC5AC),]

BE.expression.MUC5AC <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = MUC5AC, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$MUC5AC[for.printing.tmp$Tissue %in% i], .999))) +  
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.expression.MUC5AC)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Eso_MUC5AC.tiff", BE.expression.MUC5AC, device = "tiff", units="in", width=6, height=5)


for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$FABP1),]

BE.expression.FABP1 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = FABP1, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$FABP1[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(BE.expression.FABP1)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Eso_FABP1.tiff", BE.expression.FABP1, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$GPA33),]

BE.expression.GPA33 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = GPA33, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$GPA33[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(BE.expression.GPA33)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Eso_GPA33.tiff", BE.expression.GPA33, device = "tiff", units="in", width=6, height=5)


i <- c("GIM", "CIM", "NAG", "CAG")
ii <- "Gastric"
for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastic tissues

IM.tissues <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("IM")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(IM.tissues)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_tissue.tiff", IM.tissues, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$MUC5AC),]

IM.expression.MUC5AC <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = MUC5AC, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$MUC5AC[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(IM.expression.MUC5AC)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_MUC5AC.tiff", IM.expression.MUC5AC, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$FABP1),]

IM.expression.FABP1 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = FABP1, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$FABP1[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(IM.expression.FABP1)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_FABP1.tiff", IM.expression.FABP1, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$GPA33),]

IM.expression.GPA33 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = GPA33, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$GPA33[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(IM.expression.GPA33)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_GPA33.tiff", IM.expression.GPA33, device = "tiff", units="in", width=6, height=5)



i <- c("ND", "Ileum")
ii <- "Intestine"
for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastic tissues

Intestine.tissues <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestine")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(Intestine.tissues)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestine_tissue.tiff", Intestine.tissues, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$MUC5AC),]

Intestine.expression.MUC5AC <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = MUC5AC, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B")+#, limits = c(0,quantile(for.printing.tmp$MUC5AC[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestine")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(Intestine.expression.MUC5AC)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestine_MUC5AC.tiff", Intestine.expression.MUC5AC, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$FABP1),]

Intestine.expression.FABP1 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = FABP1, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$FABP1[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestine")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(Intestine.expression.FABP1)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestine_FABP1.tiff", Intestine.expression.FABP1, device = "tiff", units="in", width=6, height=5)


for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$GPA33),]

Intestine.expression.GPA33 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = GPA33, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,quantile(for.printing.tmp$GPA33[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestine")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(Intestine.expression.GPA33)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestine_GPA33.tiff", Intestine.expression.GPA33, device = "tiff", units="in", width=6, height=5)


i <- c("NG", "NGB")
ii <- "Stomach"
for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastic tissues

Stomach.tissues <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
  scale_color_manual(values = colour.palet) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Stomach")+ 
  theme_void() +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(Stomach.tissues)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Stomach_tissue.tiff", Stomach.tissues, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$MUC5AC),]

Stomach.expression.MUC5AC <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = MUC5AC, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B")+#, limits = c(0,quantile(for.printing.tmp$MUC5AC[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Stomach")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(Stomach.expression.MUC5AC)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Stomach_MUC5AC.tiff", Stomach.expression.MUC5AC, device = "tiff", units="in", width=6, height=5)

for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$FABP1),]

Stomach.expression.FABP1 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = FABP1, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B") + #, limits = c(0,quantile(for.printing.tmp$FABP1[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Stomach")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(Stomach.expression.FABP1)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Stomach_FABP1.tiff", Stomach.expression.FABP1, device = "tiff", units="in", width=6, height=5)


for.printing.tmp<-for.printing.tmp[order(for.printing.tmp$GPA33),]

Stomach.expression.GPA33 <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  
  
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = GPA33, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B") + #, limits = c(0,quantile(for.printing.tmp$FABP1[for.printing.tmp$Tissue %in% i], .999))) + 
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Stomach")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)

plot(Stomach.expression.GPA33)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Stomach_GPA33.tiff", Stomach.expression.GPA33, device = "tiff", units="in", width=6, height=5)







cell.plot<-ggplot(data.frame(UMAP1 = umap.data[,1],
                             UMAP2 = umap.data[,2],
                             Tissue = factor(sce$Celltypes_global, levels = names(cell.order)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue), size = 0.5) + 
  theme_void() +
  ylim(umap.min.max) +
  xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  # scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
  scale_color_manual(values = cell.order) 

print(cell.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Allumap_cells.tiff", cell.plot, device = "tiff", units="in", width=12, height=10)

cvdPlot(cell.plot)


# Visualise individual genes
gene1<-"KLF4"
expr.order<-order(logcounts(sce)[rowData(sce)$Symbol == gene1,])
print(ggplot(data.frame(UMAP1 = umap.data[expr.order,1],
                        UMAP2 = umap.data[expr.order,2],
                        gene = logcounts(sce)[rowData(sce)$Symbol == gene1,expr.order])) + 
        geom_point(aes(UMAP1, UMAP2, colour = gene)) + scale_colour_viridis(name = gene1, option = "B")+ 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max)
)
# Study
print(ggplot(data.frame(UMAP1 = umap.data[,1],
                        UMAP2 = umap.data[,2],
                        Study = as.factor(colData(sce)$Study))) + 
        geom_point(aes(UMAP1, UMAP2, colour = Study), size = 0.5) + 
        scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(12, "Set3"), brewer.pal(8, "Set2")))+ 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max)
)


# Is healthy
print(ggplot(data.frame(UMAP1 = umap.data[,1],
                        UMAP2 = umap.data[,2],
                        Status = as.factor(colData(sce)$Patient_type))) + 
        geom_point(aes(UMAP1, UMAP2, colour = Status), size = 0.5) + 
        scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(12, "Set3"), brewer.pal(8, "Set2")))+ 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max)
)

# Is healthy and study
print(ggplot(data.frame(UMAP1 = umap.data[,1],
                        UMAP2 = umap.data[,2],
                        Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)))) + 
        geom_point(aes(UMAP1, UMAP2, colour = Combined), size = 0.5) + 
        scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(12, "Set3"), brewer.pal(8, "Set2")))+ 
        theme_cowplot(12) +
        ylim(umap.min.max) +
        xlim(umap.min.max)
)

# Plot all patients individually 
# Generate patient colour vector
patient_vector <- c(brewer.pal(9, "Set1"), brewer.pal(12, "Set3"), brewer.pal(8, "Set2"))[1:length(unique(sce$Patient))]
names(patient_vector) <- unique(sce$Patient)

for.printing <- data.frame(UMAP1 = umap.data[,1],
                           UMAP2 = umap.data[,2],
                           Tissue = factor(sce$Tissue, levels = names(colour.palet)),
                           Patient = as.factor(sce$Patient),
                           Study = as.factor(sce$Study),
                           Time = as.factor(sce$Time),
                           CellType = factor(sce$Celltypes_global, levels = names(cell.order)),
                           Tissuetype = as.factor(sce$Tissuetypes_global),
                           Patientstatus = as.factor(sce$Patient_status),
                           Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)

qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))


for(i in unique(sce$Tissue)) {
  # for(i in c("NE", "NSCJ", "BSCJ", "BE", "NG", "ND", "SMG","GIM", "CIM", "GM","NAG", "CAG")){
  # i <- "ND"
  
  
  for.printing.tmp<-for.printing[c(which(for.printing$Tissue != i),which(for.printing$Tissue == i)), ]  
  for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue == i, i, "other")
  # for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Time == i & for.printing.tmp$Time == "0h", "other", i)
  # for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue == i & for.printing.tmp$Time == "0h" & for.printing.tmp$Study == "Sanger", i, "other")
  # 
  # # Visualise CALD1 genes
  # 
  # print(
  #   ggplot(data = for.printing.tmp) +
  #     geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
  #     geom_point(aes(UMAP1, UMAP2, colour = gene, alpha = Alpha), size = 0.5) +
  #     scale_colour_viridis(name = gene1, option = "B") + 
  #     scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
  #     ggtitle(i)+ 
  #     theme_cowplot(12) +
  #     ylim(c(floor(min(reducedDims(sce)$umap[,1:2])), ceiling(max(reducedDims(sce)$umap[,1:2])))) +
  #     xlim(c(floor(min(reducedDims(sce)$umap[,1:2])), ceiling(max(reducedDims(sce)$umap[,1:2]))))
  # )
  # 
  # Visualize Tissue
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = colour.palet) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  # Visualize  Tissue type
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = Tissuetype, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = col_vector) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  # Visualize Patient
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(data = for.printing.tmp[for.printing.tmp$Tissue == i,],
                 mapping = aes(UMAP1, UMAP2, colour = Patient), size = 0.5) +
      scale_color_manual(values = col_vector) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  
  # Visualize Cell type
  celltype.individual.plot <- ggplot(data = for.printing.tmp) +
    geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
    geom_point(aes(UMAP1, UMAP2, colour = CellType, alpha = Alpha), size = 0.5) +
    scale_color_manual(values = cell.order) + 
    scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
    ggtitle(i)+ 
    theme_void() +
    guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
    ylim(umap.min.max) +
    xlim(umap.min.max)
  
  print(celltype.individual.plot)
  ggsave(paste0("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/",i,"_umap_cells.tiff"), celltype.individual.plot, device = "tiff", units="in", width=12, height=10)
  
  # Visualize Study
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = Study, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = col_vector) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  # Visualize Patient Status
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = Patientstatus, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = col_vector) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  
  
}




for(i in levels(for.printing$Patient)){
  # i <- "Patient22"
  
  for.printing.tmp<-for.printing[c(which(for.printing$Patient != i),which(for.printing$Patient == i)), ]  
  for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Patient == i, i, "other")
  
  
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Patient != i)), rep(1, times = sum(for.printing.tmp$Patient == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = colour.palet) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
  
  # Visualize Cell type
  print(
    ggplot(data = for.printing.tmp) +
      geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = c(rep(0.01, times = sum(for.printing.tmp$Tissue != i)), rep(1, times = sum(for.printing.tmp$Tissue == i)))) +
      geom_point(aes(UMAP1, UMAP2, colour = CellType, alpha = Alpha), size = 0.5) +
      scale_color_manual(values = col_vector) + 
      scale_alpha_manual(values = c(i = 1 , "other" = 0.00)) +
      ggtitle(i)+ 
      theme_cowplot(12) +
      ylim(umap.min.max) +
      xlim(umap.min.max)
  )
}












print(table(colData(sce)$Patient, colData(sce)$Tissue))

print(colSums(table(colData(sce)$Patient, colData(sce)$Tissue)))

print(table(sce$Global_cluster_selected, colData(sce)$Tissue))

print(table(sce$Global_cluster_selected, clusters_selected))

print(table(sce$Tissue, clusters_selected))

print(table(sce$Celltypes_global, clusters_selected))

print(table(sce$Celltypes_global, sce$Tissue))

```

### Plot scatterplot

```{r}
for.printing <- data.frame(UMAP1 = umap.data[,1],
                           UMAP2 = umap.data[,2],
                           Tissue = factor(sce$Tissue, levels = names(colour.palet)),
                           # Patient = as.factor(sce$Patient),
                           # Study = as.factor(sce$Study),
                           # Time = as.factor(sce$Time),
                           CellType = factor(sce$Celltypes_global, levels = names(cell.order)),
                           MUC5AC = logcounts(sce)[rowData(sce)$Symbol == "MUC5AC",],
                           FABP1 = logcounts(sce)[rowData(sce)$Symbol == "FABP1",],
                           GPA33 = logcounts(sce)[rowData(sce)$Symbol == "GPA33",]
                           # Tissuetype = as.factor(sce$Tissuetypes_global),
                           # Patientstatus = as.factor(sce$Patient_status),
                           # Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           # Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           # Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)
for.printing.tmp <- for.printing[for.printing$Tissue %in% c("NG", "NGB", "BE", "BSCJ", "GIM", "CIM", "ND", "Ileum" ),]
for.printing.tmp$Tissue <- factor(for.printing.tmp$Tissue, levels = c("NG", "NGB", "BE", "BSCJ", "CIM", "GIM", "ND", "Ileum" ))
scatter.plot.IM <- ggplot(data = for.printing.tmp, aes(x = MUC5AC, y = GPA33, color = Tissue)) + 
  geom_point() +
  facet_wrap(vars(Tissue), ncol = 2) +
  scale_color_manual(values = colour.palet) +
  theme_cowplot(12)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Scatter_tissue.pdf", scatter.plot.IM, device = "pdf", units="in", width=6, height=10, useDingbats = FALSE)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Scatter_tissue.tiff", scatter.plot.IM, device = "tiff", units="in", width=6, height=7.5, bg = "white")


```

# Gastro-Intestinal (GI) Index

I noticed that some cell types in BE, CIM, GIM, CAG, NAG and GM look like a mixture of two cell types (e.g. they express both MUC5AC and FAPB2). Here I attempt to deconvolute the contribution of each cell type to see if we can get a score of how "Intestinal or Gastric" they are. 

## SingleR - Training

I am using my existing data for training. I will use cell labels that I already have to assign them to know cell types. Training will be done on SMG, NG, NE and ND and I will run it on all other cells. 


```{r}
# Keep only data for NE, SMG, ND, NG
sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")

sce.normal <- sce[, sce$Tissue %in% c("NGB", "ND", "NG", "Ileum")]


# run Training module of SingleR. Keep around 20 genes per cell type
SingleR.trained <- trainSingleR(ref=sce.normal, labels=sce.normal$Celltypes_global, de.method="wilcox", de.n = 20)

# Get expression data in the TPM format
tpm(sce.normal) <- calculateTPM(sce.normal)

# Costruct tibble with the required data
sce.matrix <- as_tibble(t(as.matrix(tpm(sce.normal))))
sce.matrix$label<-sce.normal$Celltypes_global


# Test the expression of markers
Marker_heatmap(sce.matrix, SingleR.trained$common.genes)
```

It looks like the markers are quite specific.


## Annotate the cells

First, I will do it on normal cells

```{r, eval = FALSE}
library(BiocParallel)
param <- MulticoreParam(workers = 40) # run with all cores

# Run SingleR with single cell option and and use wilcox method for DE
prd.SingleR.BE <- classifySingleR(test=sce.normal , trained=SingleR.trained, BPPARAM = param)
# prd.SingleR.BE <- SingleR(test=sce.BE.test , ref=sce.BE, labels=sce.BE$cell_type_secondary, de.method="wilcox", BPPARAM = param)

```

## Visualise cell types

### General overview

```{r}
# Print number of cells assigned per cell type
table(prd.SingleR.BE$labels)

# Print number of cells assigned per cell type per tissue type
cell.order.ref<-names(cell.order)[names(cell.order) %in% unique(sce.normal$Celltypes_global)]
prd.SingleR.BE$pruned.labels <- factor(prd.SingleR.BE$pruned.labels, levels = cell.order.ref)
prd.SingleR.BE$labels <- factor(prd.SingleR.BE$labels, levels = cell.order.ref)
prd.SingleR.BE$first.labels <- factor(prd.SingleR.BE$first.labels, levels = cell.order.ref)
prd.SingleR.BE$scores <- prd.SingleR.BE$scores[,cell.order.ref]


print(table(prd.SingleR.BE$pruned.labels, sce.normal$Tissue, useNA = "ifany"))

# print the pruned cell types
print(table(prd.SingleR.BE$labels[is.na(prd.SingleR.BE$pruned.labels)], colData(sce.normal)$Tissue[is.na(prd.SingleR.BE$pruned.labels)], useNA = "ifany"))

# More stringent pruning (nmads2)
cell.labels<-prd.SingleR.BE$pruned.labels
cell.labels[pruneScores(prd.SingleR.BE, nmads = 2)] <- NA
print(table(prd.SingleR.BE$labels[is.na(cell.labels)], colData(sce.normal)$Tissue[is.na(cell.labels)], useNA = "ifany"))

# Test the accuracy of identified cells

# fix he confusion_heatmap script from scriBet package to sort the cells by levels of factors inputed into the function
.Confusion_heatmap <- function (ori, prd, order.ori, order.prd) {
  cross.validation.filt <- tibble(ori = ori, prd = prd) %>% 
    dplyr::count(ori, prd) %>% tidyr::spread(key = prd, value = n)
  cross.validation.filt[is.na(cross.validation.filt)] = 0
  cross.validation.filt[, -1] <- round(cross.validation.filt[, 
                                                             -1]/rowSums(cross.validation.filt[, -1]), 2)
  cross.validation.filt <- cross.validation.filt %>% tidyr::gather(key = "prd", 
                                                                   value = "Prob", -ori)
  cross.validation.filt$ori <- factor(cross.validation.filt$ori , levels = order.ori)
  cross.validation.filt$prd <- factor(cross.validation.filt$prd , levels = order.prd)
  p <- cross.validation.filt %>% ggplot(aes(ori, prd, fill = Prob)) + 
    geom_tile() + theme(axis.title = element_text(size = 0)) + 
    theme(axis.text = element_text(size = 10)) + theme(legend.title = element_text(size = 0)) + 
    theme(legend.text = element_text(size = 10)) + theme(panel.grid.major = element_blank(), 
                                                         panel.grid.minor = element_blank(), panel.background = element_blank(), 
                                                         axis.ticks = element_blank(), axis.title = element_blank()) + 
    theme(axis.text.y = element_text(color = "black"), axis.text.x = element_text(color = "black", 
                                                                                  angle = 45, hjust = 1)) + scale_fill_viridis()
  return(p)
}

.Confusion_heatmap(sce.normal$Celltypes_global, prd.SingleR.BE$labels, order.ori = cell.order.ref, order.prd = cell.order.ref )
.Confusion_heatmap(sce.normal$Celltypes_global, prd.SingleR.BE$pruned.labels, order.ori = cell.order.ref, order.prd = c(cell.order.ref, "NA"))



# Plot score distrubution and how it was affected by prunning
plotScoreDistribution(prd.SingleR.BE)
# plotScoreDistribution(prd.SingleR.BE, show = "scores")

```


## Annotate all cells

Let's do it again on all cells

```{r, eval = FALSE}
library(BiocParallel)
param <- MulticoreParam(workers = 40) # run with all cores

# Run SingleR with single cell option and and use wilcox method for DE
prd.SingleR.All <- classifySingleR(test=sce , trained=SingleR.trained, BPPARAM = param)
# prd.SingleR.BE <- SingleR(test=sce.BE.test , ref=sce.BE, labels=sce.BE$cell_type_secondary, de.method="wilcox", BPPARAM = param)

```

## Visualise cell types

### General overview

```{r}
# Print number of cells assigned per cell type
table(prd.SingleR.All$labels)

# Print number of cells assigned per cell type per tissue type
cell.order.ref<-names(cell.order)[names(cell.order) %in% unique(sce$Celltypes_global)]
prd.SingleR.All$pruned.labels <- factor(prd.SingleR.All$pruned.labels, levels = cell.order.ref)
prd.SingleR.All$labels <- factor(prd.SingleR.All$labels, levels = cell.order.ref)
prd.SingleR.All$first.labels <- factor(prd.SingleR.All$first.labels, levels = cell.order.ref)
prd.SingleR.All$scores <- prd.SingleR.All$scores[,cell.order.ref]


print(table(prd.SingleR.All$pruned.labels, sce$Tissue, useNA = "ifany"))

# print the pruned cell types
print(table(prd.SingleR.All$labels[is.na(prd.SingleR.All$pruned.labels)], colData(sce)$Tissue[is.na(prd.SingleR.All$pruned.labels)], useNA = "ifany"))

# More stringent pruning (nmads2)
cell.labels<-prd.SingleR.All$pruned.labels
cell.labels[pruneScores(prd.SingleR.All, nmads = 2)] <- NA
print(table(prd.SingleR.All$labels[is.na(cell.labels)], colData(sce)$Tissue[is.na(cell.labels)], useNA = "ifany"))

# Test the accuracy of identified cells

# fix he confusion_heatmap script from scriBet package to sort the cells by levels of factors inputed into the function
.Confusion_heatmap <- function (ori, prd, order.ori, order.prd) {
  cross.validation.filt <- tibble(ori = ori, prd = prd) %>% 
    dplyr::count(ori, prd) %>% tidyr::spread(key = prd, value = n)
  cross.validation.filt[is.na(cross.validation.filt)] = 0
  cross.validation.filt[, -1] <- round(cross.validation.filt[, 
                                                             -1]/rowSums(cross.validation.filt[, -1]), 2)
  cross.validation.filt <- cross.validation.filt %>% tidyr::gather(key = "prd", 
                                                                   value = "Prob", -ori)
  cross.validation.filt$ori <- factor(cross.validation.filt$ori , levels = order.ori)
  #count the proportion of lowest prediction
  lowest.predict<-aggregate(x = cross.validation.filt$Prob, by = list(cross.validation.filt$prd), FUN = max)
  cross.validation.filt <- cross.validation.filt[cross.validation.filt$prd %in% lowest.predict[lowest.predict[,2]>0.01,1],]
  cross.validation.filt$prd <- factor(cross.validation.filt$prd , levels = order.prd[order.prd %in% unique(cross.validation.filt$prd)])
  p <- cross.validation.filt %>% ggplot(aes(ori, prd, fill = Prob)) + 
    geom_tile() + theme(axis.title = element_text(size = 0)) + 
    theme(axis.text = element_text(size = 10)) + theme(legend.title = element_text(size = 0)) + 
    theme(legend.text = element_text(size = 10)) + theme(panel.grid.major = element_blank(), 
                                                         panel.grid.minor = element_blank(), panel.background = element_blank(), 
                                                         axis.ticks = element_blank(), axis.title = element_blank()) + 
    theme(axis.text.y = element_text(color = "black"), axis.text.x = element_text(color = "black", 
                                                                                  angle = 45, hjust = 1)) + scale_fill_viridis(option="turbo",limits=c(0, 1))
  return(p)
}

print.tissue<-lapply(
  unique(sce$Tissue), 
  function (x) 
    .Confusion_heatmap(
      sce$Celltypes_global[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]], 
      prd.SingleR.All$labels[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]], 
      order.ori = cell.order.ref[cell.order.ref %in% unique(sce$Celltypes_global[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]])], 
      order.prd = cell.order.ref[cell.order.ref %in% unique(prd.SingleR.All$labels[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]])]) + 
    labs(title = x)
)
names(print.tissue)<-unique(sce$Tissue)
gridExtra::grid.arrange(grobs = print.tissue[1:4], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue[5:8], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue[9:12], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue[13:14], ncol = 2, name = "test")


print.tissue.pruned<-lapply(
  unique(sce$Tissue), 
  function (x) 
    .Confusion_heatmap(
      sce$Celltypes_global[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]], 
      prd.SingleR.All$pruned.labels[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]], 
      order.ori = cell.order.ref[cell.order.ref %in% unique(sce$Celltypes_global[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]])], 
      order.prd = c(cell.order.ref[cell.order.ref %in% unique(prd.SingleR.All$pruned.labels[sce$Tissue == x & sce$Celltypes_global %in% names(table(sce$Celltypes_global[sce$Tissue == x]))[table(sce$Celltypes_global[sce$Tissue == x])>20]])], "NA")) + 
    labs(title = x)# +  scale_fill_continuous(limits=c(0, 1))
)
names(print.tissue.pruned)<-unique(sce$Tissue)
gridExtra::grid.arrange(grobs = print.tissue.pruned[1:4], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue.pruned[5:8], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue.pruned[9:12], ncol = 2, name = "test")
gridExtra::grid.arrange(grobs = print.tissue.pruned[13:14], ncol = 2, name = "test")

print.tissue.pruned<-print.tissue.pruned[match(c("ND", "Ileum", "Colon", "Rectum", "NG", "NSCJ", "NGB", "GM", "BE", "BSCJ", "NAG", "CAG", "GIM", "CIM"), names(print.tissue.pruned))]
print.all<-gridExtra::marrangeGrob(grobs = print.tissue.pruned, ncol = 2 , nrow = 2)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/prediction.pdf", print.all, device = "pdf", units="in", width=15, height=13)


# Plot score distrubution and how it was affected by prunning
plotScoreDistribution(prd.SingleR.All)

```

# MuSiC deconvolution


## Run MuSiC, no intermediate with SMG and NE

```{r, eval = TRUE}
sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_after_exclusion2.rds")
sce<-sce[,sce$Tissuetypes_global %in% c("Columnar", "Glandular", "Squamous")]
sce.normal <- sce[, (sce$Tissue %in% c("ND", "Ileum", "Colon") & sce$Celltypes_global %in% c("Enterocytes", "Enterocytes_Intermediate", "Intestinal_Undifferentiated", "Goblet", "Enteroendocrine_CHGA" )) |
                    (sce$Tissue %in% c("NGB", "NG") & sce$Celltypes_global %in% c("Foveolar_Differentiated", "Foveolar_Intermediate", "Neck-Cells", "Enteroendocrine_CHGA", "Enteroendocrine_GHRL", "Parietal", "Chief" )) |
                    (sce$Tissue %in% c("SMG") & sce$Celltypes_global %in% c("SMG" )) |
                    (sce$Tissue %in% c("NE") & sce$Celltypes_global %in% c("Basal", "Suprabasal", "Intermediate", "Superficial")) |
                    (sce$Tissue %in% c("NSCJ") & sce$Celltypes_global %in% c("SMG" ))
]



sce.normal$Celltypes_global[sce.normal$Celltypes_global %in% c("Foveolar_Differentiated", "Foveolar_Intermediate")]<-"Foveolar"
sce.normal$Celltypes_global[sce.normal$Celltypes_global %in% c("Enterocytes_Intermediate", "Enterocytes")]<-"Enterocytes"

#create phenotype matrix
pheno.matrix<-colData(sce.normal)[,c("Sample", "Patient", "Tissue", "Celltypes_global")]
pheno.matrix$Celltype.tissue <- pheno.matrix$Tissue

pheno.matrix$Celltype.tissue[pheno.matrix$Tissue %in%  c("ND", "Ileum")] <- "Intestine"
pheno.matrix$Celltype.tissue[pheno.matrix$Tissue %in%  c("NG", "NGB")] <- "Gastric"
# pheno.matrix$Celltype.tissue[pheno.matrix$Tissue == "Colon"] <- "Colon"
pheno.matrix$Celltype.tissue<-paste(pheno.matrix$Celltype.tissue, pheno.matrix$Celltypes_global, sep = "_")
#annotate phenotypes
metadata <- data.frame(labelDescription= c("Sample", "Patient", "Tissue", "Cell Type", "Cell Type Combined"), row.names=colnames(pheno.matrix))

#create Expression set for the input into MuSiC
cur.sce1<-ExpressionSet(assayData = as.matrix(counts(sce.normal)), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix), varMetadata = metadata) )


# Get data for the test samples, all data is contained in sce
# sce.BE.test.MuSiC<-sce
# sce.BE.test.MuSiC<-sce[,sce$Celltypes_global %in% c("Enterocytes", "Enterocytes_Intermediate", "Foveolar_Differentiated", "Foveolar_Intermediate", "Intestinal_Undifferentiated", "Neck-Cells" )]
# jumbled_MuSiC <- sample(ncol(sce.BE.test.MuSiC))
# jumbled_MuSiC <- readRDS("~/Dropbox/Postdoc/2019-10-29_Gastric_IM/Results/MuSiC_randomno.RDS")

n.threat<-24 #number of CPU
ncells<-ceiling(seq(0, ncol(sce), length.out = n.threat+1))  # number of cells per CPU to be analysed
# Around 4.5K cells
# ncells <- 5000

# sce.BE.test.MuSiC<-sce.BE.test.MuSiC[,jumbled_MuSiC[1:ncells]]
pheno.matrix.test <-colData(sce)[,c("Sample", "Patient", "Tissue", "Celltypes_global")]

metadata2<-data.frame(labelDescription= c("Sample", "Patient", "Tissue", "Celltypes_global"), row.names=colnames(pheno.matrix.test))


```

```{r, eval = FALSE}
# Run MuSiC in parallel
# Compute tsne on corrected counts
set.seed(50014)

# tsne <- Rtsne(t(corrected), pca = FALSE)

# Split single cell object into small files and save them as individual .Rdata files. See option 3 here: https://lcolladotor.github.io/2013/11/14/reducing-memory-overhead-when-using-mclapply/#.YPbxgOy8g-U
# it makes the memory footprint smallest in principle
for (i in c(1:n.threat)) {
  data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce[,(ncells[i]+1):(ncells[i+1])])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):(ncells[i+1]),]), varMetadata = metadata2))
  save(data.all.test, file = paste0("chunk", i, ".Rdata"))
}

rm(sce)
rm(sce.normal)
gc()

run.music<-function(x) {
  # print(i)
  load(paste0("chunk", x, ".Rdata"))
  cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltype.tissue", samples = "Sample", select.ct = unique(cur.sce1$Celltype.tissue), verbose = FALSE)
  return(cell.prop)
}


results.no.intermediate<-mclapply(1:n.threat, 
                                  FUN = run.music, 
                                  mc.cores = 2)
for (i in c(1:n.threat)) {
  fn<-paste0("chunk", i, ".Rdata")
  if (file.exists(fn)) {
    #Delete file if it exists
    file.remove(fn)
  }
}


# cl <- makeCluster(2) #not to overload your computer
# registerDoParallel(cl)

# results.no.intermediate <-  foreach(i = c(1:n.threat)) %dopar% {
# results.no.intermediate <-  foreach(i = c(1:n.threat)) %do% {
#   library(MuSiC)
#   library(scran)
#   #create expression set
#   # data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce.BE.test.MuSiC[,(ncells[i]+1):10])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):10,]), varMetadata = metadata2))
#   data.all.test<-ExpressionSet(assayData = as.matrix(counts(sce[,(ncells[i]+1):(ncells[i+1])])), phenoData = new("AnnotatedDataFrame", data = as.data.frame(pheno.matrix.test[(ncells[i]+1):(ncells[i+1]),]), varMetadata = metadata2))
#   cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltype.tissue", samples = "Sample", select.ct = unique(cur.sce1$Celltype.tissue), verbose = FALSE)
#   #do other things if you want
#   return(cell.prop) 
# }
# stopCluster(cl)
estimated.weights.no.intermediate<-do.call(rbind,lapply(results.no.intermediate, function (x) return(x$Est.prop.weighted)))
# cell.prop<-music_prop(bulk.eset = data.all.test, sc.eset = cur.sce1, clusters = "Celltypes_global", samples = "Patient", select.ct = unique(cur.sce1$cell.types), verbose = FALSE)
# cell.prop.old<-cell.prop
# cell.prop<-cell.prop.all

saveRDS(results.no.intermediate, "~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/MuSiC.rds")
# saveRDS(jumbled_MuSiC, "~/Dropbox/Postdoc/2019-10-29_Gastric_IM/Results/MuSiC_randomno.rds")
```

## UMAP for all cells

```{r}
results.no.intermediate <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/MuSiC.rds")
estimated.weights.no.intermediate<-do.call(rbind,lapply(results.no.intermediate, function (x) return(x$Est.prop.weighted)))

sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_after_exclusion2.rds")
sce<-sce[,sce$Tissuetypes_global %in% c("Columnar", "Glandular", "Squamous")]
umap.data<-reducedDims(sce)$umap_MinDist_0.2_N_Neighbors_15 
umap.min.max<-c(floor(min(umap.data[,1:2])), ceiling(max(umap.data[,1:2])))




for.printing <- data.frame(UMAP1 = umap.data[,1],
                           UMAP2 = umap.data[,2],
                           Tissue = factor(sce$Tissue, levels = names(colour.palet)),
                           # Patient = as.factor(sce$Patient),
                           # Study = as.factor(sce$Study),
                           # Time = as.factor(sce$Time),
                           CellType = factor(sce$Celltypes_global, levels = names(cell.order)),
                           NE = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("NE", colnames(estimated.weights.no.intermediate))]), 
                           SMG = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("SMG_SMG", colnames(estimated.weights.no.intermediate)), drop = FALSE]), 
                           Transitional = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("NSCJ_SMG", colnames(estimated.weights.no.intermediate)), drop = FALSE]), 
                           Colon = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Colon", colnames(estimated.weights.no.intermediate))]), 
                           Gastric = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Gastric", colnames(estimated.weights.no.intermediate))]), 
                           Intestine = matrixStats::rowSums2(estimated.weights.no.intermediate[,grep("Intestine", colnames(estimated.weights.no.intermediate))]), 
                           MUC5AC = logcounts(sce)[rowData(sce)$Symbol == "MUC5AC",],
                           FABP1 = logcounts(sce)[rowData(sce)$Symbol == "FABP1",]
                           # Tissuetype = as.factor(sce$Tissuetypes_global),
                           # Patientstatus = as.factor(sce$Patient_status),
                           # Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           # Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           # Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)


set.seed(12345)
jumbled<-sample(nrow(for.printing), replace = FALSE)
for.printing<-for.printing[jumbled,]

i <- unique(for.printing$Tissue)

for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Esophageal tissues
All.tissues.Esophageal <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = NE, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Esophageal Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.Esophageal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Esophageal_all.tiff", All.tissues.Esophageal, device = "tiff", units="in", width=6, height=5)

# Visualize Intestine tissues
All.tissues.Intestine <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Intestine, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestinal Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.Intestine)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestine_all.tiff", All.tissues.Intestine, device = "tiff", units="in", width=6, height=5)


# Visualize Colon tissues
All.tissues.Colon <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Colon, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Colonic Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.Colon)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Colon_all.tiff", All.tissues.Colon, device = "tiff", units="in", width=6, height=5)

# Visualize SMG tissues
All.tissues.SMG <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = SMG, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("SMG Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.SMG)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/SMG_all.tiff", All.tissues.SMG, device = "tiff", units="in", width=6, height=5)

# Visualize Transitional tissues
All.tissues.Transitional <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Transitional, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Transitional Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.Transitional)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Transitional_all.tiff", All.tissues.Transitional, device = "tiff", units="in", width=6, height=5)

# Visualize Transitional tissues
All.tissues.Gastric <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Gastric, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Gastric Phenotype")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(All.tissues.Gastric)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_all.tiff", All.tissues.Gastric, device = "tiff", units="in", width=6, height=5)


######################
###finished here
######################

```

## UMAP only for columnar

```{r}
sce<-readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_after_exclusion2.rds")
sce<-sce[,sce$Tissuetypes_global %in% c("Columnar", "Glandular", "Squamous")]
sce.columnar <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")
results.no.intermediate <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/MuSiC.rds")

estimated.weights.no.intermediate<-do.call(rbind,lapply(results.no.intermediate, function (x) return(x$Est.prop.weighted)))

columnar.weight<-estimated.weights.no.intermediate[match(paste(sce.columnar$Sample, sce.columnar$Barcode),paste(sce$Sample, sce$Barcode)),]

umap.data<-reducedDims(sce.columnar)$umap_MinDist_0.2_N_Neighbors_15 
umap.min.max<-c(floor(min(umap.data[,1:2])), ceiling(max(umap.data[,1:2])))




for.printing <- data.frame(UMAP1 = umap.data[,1],
                           UMAP2 = umap.data[,2],
                           Tissue = factor(sce.columnar$Tissue, levels = names(colour.palet)),
                           Patient = as.factor(sce.columnar$Patient),
                           # Study = as.factor(sce$Study),
                           # Time = as.factor(sce$Time),
                           CellType = factor(sce.columnar$Celltypes_global, levels = names(cell.order)),
                           NE = matrixStats::rowSums2(columnar.weight[,grep("NE", colnames(columnar.weight))]), 
                           SMG = matrixStats::rowSums2(columnar.weight[,grep("SMG_SMG", colnames(columnar.weight)), drop = FALSE]), 
                           Transitional = matrixStats::rowSums2(columnar.weight[,grep("NSCJ_SMG", colnames(columnar.weight)), drop = FALSE]), 
                           Colon = matrixStats::rowSums2(columnar.weight[,grep("Colon", colnames(columnar.weight))]), 
                           Gastric = matrixStats::rowSums2(columnar.weight[,grep("Gastric", colnames(columnar.weight))]), 
                           Intestine = matrixStats::rowSums2(columnar.weight[,grep("Intestine", colnames(columnar.weight))]), 
                           MUC5AC = logcounts(sce.columnar)[rowData(sce.columnar)$Symbol == "MUC5AC",],
                           FABP1 = logcounts(sce.columnar)[rowData(sce.columnar)$Symbol == "FABP1",]
                           # Tissuetype = as.factor(sce$Tissuetypes_global),
                           # Patientstatus = as.factor(sce$Patient_status),
                           # Combined = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_type)),
                           # Combined2 = as.factor(paste(colData(sce)$Study, colData(sce)$Patient_status)),
                           # Combined3 = as.factor(paste(colData(sce)$Patient, sce$Time))
)


set.seed(12345)
jumbled<-sample(nrow(for.printing), replace = FALSE)
for.printing<-for.printing[jumbled,]

i<-c("BSCJ", "BE", "GM")
ii<-"Eso"

for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastric pghenotype in BE tissues
BE.tissues.gastric <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Gastric, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Gastric Phenotype in BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues.gastric)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_in_BE.tiff", BE.tissues.gastric, device = "tiff", units="in", width=6, height=5)



# Visualize Intestianl pghenotype in BE tissues
BE.tissues.Intestinal <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Intestine, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestinal Phenotype in BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues.Intestinal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestinal_in_BE.tiff", BE.tissues.Intestinal, device = "tiff", units="in", width=6, height=5)


# Visualize Colon pghenotype in BE tissues
BE.tissues.Colon <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Colon, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Colonic Phenotype in BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues.Colon)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Colon_in_BE.tiff", BE.tissues.Colon, device = "tiff", units="in", width=6, height=5)

# Visualize Colon pghenotype in BE tissues
BE.tissues.NE <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = NE, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("NE Phenotype in BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues.NE)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/NE_in_BE.tiff", BE.tissues.NE, device = "tiff", units="in", width=6, height=5)

# Visualize Colon pghenotype in BE tissues
BE.tissues.SMG <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = SMG, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("SMG Phenotype in BE")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(BE.tissues.SMG)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/SMG_in_BE.tiff", BE.tissues.SMG, device = "tiff", units="in", width=6, height=5)



i<-c("NAG", "CAG", "GIM", "CIM")
ii<-"Gas"

for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")



# Visualize Gastric pghenotype in GIM tissues
GIM.tissues.gastric <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Gastric, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Gastric Phenotype in IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(GIM.tissues.gastric)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Gastric_in_IM.tiff", GIM.tissues.gastric, device = "tiff", units="in", width=6, height=5)

# Visualize Intestianl pghenotype in IM tissues
GIM.tissues.Intestinal <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Intestine, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestinal Phenotype in IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(GIM.tissues.Intestinal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Intestinal_in_IM.tiff", GIM.tissues.Intestinal, device = "tiff", units="in", width=6, height=5)

# Visualize Intestianl pghenotype in IM tissues
GIM.tissues.NE <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = NE, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("NE Phenotype in IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(GIM.tissues.NE)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/NE_in_IM.tiff", GIM.tissues.NE, device = "tiff", units="in", width=6, height=5)

# Visualize Intestianl pghenotype in IM tissues
GIM.tissues.Colon <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Colon, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Colonic Phenotype in IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(GIM.tissues.Colon)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Colon_in_IM.tiff", GIM.tissues.Colon, device = "tiff", units="in", width=6, height=5)

# Visualize Intestianl pghenotype in IM tissues
GIM.tissues.SMG <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = SMG, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("SMG Phenotype in IM")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(GIM.tissues.SMG)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/SMG_in_IM.tiff", GIM.tissues.SMG, device = "tiff", units="in", width=6, height=5)




```

# Normal

```{r}
i<-c("NG", "NGB")
ii<-"Stomach"

for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastric pghenotype in BE tissues
normalstomach.tissues.gastric <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Gastric, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Gastric Phenotype in Stomach")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(normalstomach.tissues.gastric)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Contributions_normal/Gastric_in_Stomach.tiff", normalstomach.tissues.gastric, device = "tiff", units="in", width=6, height=5, bg = "white")



# Visualize Intestianl pghenotype in BE tissues
normalstomach.tissues.Intestinal <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Intestine, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestinal Phenotype in Stomach")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(normalstomach.tissues.Intestinal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Contributions_normal/Intestinal_in_Stomach.tiff", normalstomach.tissues.Intestinal, device = "tiff", units="in", width=6, height=5, bg = "white")

i<-c("ND", "Ileum")
ii<-"Intestine"

for.printing.tmp<-for.printing[c(which(!(for.printing$Tissue %in% i)),which(for.printing$Tissue %in% i)), ]  
for.printing.tmp$Alpha <- ifelse(for.printing.tmp$Tissue %in% i, ii, "other")

# Visualize Gastric pghenotype in BE tissues
normalintestine.tissues.gastric <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Gastric, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Gastric Phenotype in Inestine")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(normalintestine.tissues.gastric)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Contributions_normal/Gastric_in_Intestine.tiff", normalintestine.tissues.gastric, device = "tiff", units="in", width=6, height=5, bg = "white")



# Visualize Intestianl pghenotype in BE tissues
normalIntestine.tissues.Intestinal <- ggplot(data = for.printing.tmp) +
  geom_point(aes(UMAP1, UMAP2, alpha = Alpha), size = 1, color = "black", alpha = ifelse(for.printing.tmp$Tissue %in% i,1,0.01))+
  # c(rep(0.01, times = sum(!(for.printing.tmp$Tissue %in% i))), rep(1, times = sum(for.printing.tmp$Tissue %in% i)))) +
  geom_point(aes(UMAP1, UMAP2, colour = Intestine, alpha = Alpha), size = 0.5) +
  scale_colour_viridis(option = "B", limits = c(0,1)) +
  scale_alpha_manual(values = c(ii = 1 , "other" = 0.00)) +
  ggtitle("Intestinal Phenotype in Intestine")+ 
  theme_void() +
  # guides(color=guide_legend(ncol=1, override.aes = list(size=4))) +
  ylim(umap.min.max) +
  xlim(umap.min.max)
plot(normalIntestine.tissues.Intestinal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Contributions_normal/Intestinal_in_Intestine.tiff", normalIntestine.tissues.Intestinal, device = "tiff", units="in", width=6, height=5, bg = "white")


```
## Violin plots

```{r}

for.printing <- data.frame(
  Tissue = factor(sce.columnar$Tissue, levels = names(colour.palet)),
  CellType = factor(sce.columnar$Celltypes_global, levels = names(cell.order)),
  Squamous = matrixStats::rowSums2(columnar.weight[,grep("NE", colnames(columnar.weight))]), 
  SMG = matrixStats::rowSums2(columnar.weight[,grep("SMG_SMG", colnames(columnar.weight)), drop = FALSE]), 
  Transitional = matrixStats::rowSums2(columnar.weight[,grep("NSCJ_SMG", colnames(columnar.weight)), drop = FALSE]), 
  Colon = matrixStats::rowSums2(columnar.weight[,grep("Colon", colnames(columnar.weight))]), 
  Gastric = matrixStats::rowSums2(columnar.weight[,grep("Gastric", colnames(columnar.weight))]), 
  Intestine = matrixStats::rowSums2(columnar.weight[,grep("Intestine", colnames(columnar.weight))])
  
)


for.printing.long<-reshape2::melt(for.printing,variable.name = "Contributor", value.name = "Contribution" )
for.printing.long$Contributor <- factor(for.printing.long$Contributor, levels = c("Squamous", "SMG", "Transitional", "Gastric", "Intestine", "Colon"))
for.printing.long$Tissue <- factor(for.printing.long$Tissue, levels = names(colour.palet))
for.printing.long$Tissue_type <- NA
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NSCJ", "NG", "NGB")] <- "Stomach"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("ND", "Ileum", "Colon", "Rectum")] <- "Intestine"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("BE", "BSCJ", "GM")] <- "Esophageal IM"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NAG", "CAG", "GIM", "CIM")] <- "Gastric IM"
for.printing.long$Tissue_type <- factor(for.printing.long$Tissue_type, levels = c("Stomach", "Intestine", "Esophageal IM", "Gastric IM"))
colour.palet.contributor <- c("Squamous" = "#990F0F", 
                              "Transitional" = "#CC5151",
                              "SMG" = "#FFB2B2",
                              "Gastric" = "#6B990F",
                              "Intestine" = "#260F99",
                              "Colon" = "#6551CC")
Violin.plot.all <- ggplot(data = for.printing.long, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x")
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.all.pdf", Violin.plot.all, device = "pdf", units="in", width=8, height=5, useDingbats = FALSE)

for.printing.long.IM<-for.printing.long[for.printing.long$Tissue_type %in% c("Esophageal IM", "Gastric IM"),]
tmp<-for.printing.long[for.printing.long$Tissue %in% c("ND", "NG"),]
tmp$Tissue_type <- "Esophageal IM"
for.printing.long.IM<-rbind(for.printing.long.IM, tmp)
tmp$Tissue_type <- "Gastric IM"
for.printing.long.IM<-rbind(for.printing.long.IM, tmp)
for.printing.long.IM$Tissue <- factor(for.printing.long.IM$Tissue, levels = c("NG", "GM", "BE", "BSCJ", "NAG", "CAG", "GIM", "CIM", "ND"))
for.printing.long.IM<-for.printing.long.IM[for.printing.long.IM$Contributor %in% c("Squamous", "SMG", "Gastric", "Intestine"),]

Violin.plot.IM <- ggplot(data = for.printing.long.IM, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE, size = 0.1) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x", ncol = 1) #+
# coord_flip() +
# scale_x_discrete(limits=rev)
plot(Violin.plot.IM)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.IM.pdf", Violin.plot.IM, device = "pdf", units="in", width=4, height=5.5, useDingbats = FALSE)


```

# Violin per cell type

```{r}

for.printing <- data.frame(
  Tissue = factor(sce.columnar$Tissue, levels = names(colour.palet)),
  Patient = as.factor(sce.columnar$Patient),
  # Study = as.factor(sce$Study),
  # Time = as.factor(sce$Time),
  CellType = factor(sce.columnar$Celltypes_global, levels = names(cell.order)),
  
  # Tissue = factor(sce.columnar$Tissue, levels = names(colour.palet)),
  # CellType = factor(sce.columnar$Celltypes_global, levels = names(cell.order)),
  Squamous = matrixStats::rowSums2(columnar.weight[,grep("NE", colnames(columnar.weight))]), 
  SMG = matrixStats::rowSums2(columnar.weight[,grep("SMG_SMG", colnames(columnar.weight)), drop = FALSE]), 
  Transitional = matrixStats::rowSums2(columnar.weight[,grep("NSCJ_SMG", colnames(columnar.weight)), drop = FALSE]), 
  Colon = matrixStats::rowSums2(columnar.weight[,grep("Colon", colnames(columnar.weight))]), 
  Gastric = matrixStats::rowSums2(columnar.weight[,grep("Gastric", colnames(columnar.weight))]), 
  Intestine = matrixStats::rowSums2(columnar.weight[,grep("Intestine", colnames(columnar.weight))])
  
)


for.printing.long<-reshape2::melt(for.printing,variable.name = "Contributor", value.name = "Contribution" )
for.printing.long$Contributor <- factor(for.printing.long$Contributor, levels = c("Squamous", "SMG", "Transitional", "Gastric", "Intestine", "Colon"))
for.printing.long$Tissue <- factor(for.printing.long$Tissue, levels = names(colour.palet))
for.printing.long$Tissue_type <- NA
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NSCJ", "NG", "NGB")] <- "Stomach"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("ND", "Ileum", "Colon", "Rectum")] <- "Intestine"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("BE", "BSCJ", "GM")] <- "Esophageal IM"
for.printing.long$Tissue_type[for.printing.long$Tissue %in% c("NAG", "CAG", "GIM", "CIM")] <- "Gastric IM"
for.printing.long$Tissue_type <- factor(for.printing.long$Tissue_type, levels = c("Stomach", "Intestine", "Esophageal IM", "Gastric IM"))
# for.printing.long$CellType <- factor(for.printing.long$CellType, levels = names(cell.order))


colour.palet.contributor <- c("Squamous" = "#990F0F", 
                              "Transitional" = "#CC5151",
                              "SMG" = "#FFB2B2",
                              "Gastric" = "#6B990F",
                              "Intestine" = "#260F99",
                              "Colon" = "#6551CC")

# Intestine
for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Intestine" & !for.printing.long$CellType %in% c("Enteroendocrine_GHRL", "Foveolar_Intermediate", "Neck-Cells", "Chief", "Parietal", "Enteroendocrine_GAST", "Foveolar_Differentiated") ,]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(CellType), drop = TRUE,scales = "free_x")
print(Violin.plot.all)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.Intestine.per.celltype.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)

# Stomach
for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Stomach" ,]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(CellType), drop = TRUE,scales = "free_x")
print(Violin.plot.all)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.stomach.per.celltype.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)

# Stomach IM
for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Gastric IM" ,]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(CellType), drop = TRUE,scales = "free_x")
print(Violin.plot.all)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.stomachIM.per.celltype.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)

# BE IM
name.changes <-c("NE" = "NE",
                 "NSCJ" = "NSCJ",
                 "SMG" = "SMG",
                 "GM" = "E-GM",
                 "BE" = "BE-IM",
                 "BSCJ" = "BSCJ",
                 "NG" = "NGC",
                 "NGB" = "NGB",
                 "NAG" = "NAG",
                 "CAG" = "CAG",
                 "GIM" = "GIM",
                 "CIM" = "CIM",
                 "ND" = "ND",
                 "Ileum" = "Ileum", 
                 "Colon" = "Colon",
                 "Rectum" = "Rectum")

for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Esophageal IM" & !for.printing.long$CellType %in% c("Enterocytes", "Enteroendocrine_GHRL","Chief", "Parietal", "Enteroendocrine_GAST", "Foveolar_Differentiated") ,]
for.printing.long.tmp$Tissue <- name.changes[for.printing.long.tmp$Tissue]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(CellType), drop = TRUE,scales = "free_x")
print(Violin.plot.all)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.Eso-IM.per.celltype.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)

# Per Patient
for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Gastric IM" ,]
for.printing.long.tmp$Tissue <- name.changes[for.printing.long.tmp$Tissue]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Patient), drop = TRUE,scales = "free_x")
print(Violin.plot.all)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.Stomach-IM.per.patient.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)

# Per Patient
for.printing.long.tmp <- for.printing.long[for.printing.long$Tissue_type == "Esophageal IM" & !for.printing.long$CellType %in% c("Enterocytes", "Enteroendocrine_GHRL","Chief", "Parietal", "Enteroendocrine_GAST", "Foveolar_Differentiated") ,]
for.printing.long.tmp$Tissue <- name.changes[for.printing.long.tmp$Tissue]
Violin.plot.all <- ggplot(data = for.printing.long.tmp, aes(x = Tissue, y = Contribution)) +
  theme_cowplot(12) +
  ylim(0,1) +
  geom_violin(aes(fill = Contributor), scale = "width", trim = TRUE) + 
  scale_fill_manual(values = colour.palet.contributor) +
  facet_wrap(vars(Patient), drop = TRUE,scales = "free_x")
print(Violin.plot.all)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Violin.plot.Eso-IM.per.patient.pdf", Violin.plot.all, device = "pdf", units="in", width=11, height=7, useDingbats = FALSE)


```
# Trajectory

## All cells

```{r}
library(monocle3)
# Load data 
sce <- readRDS("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Final/All_sce_selected_corrected_Columnar.rds")
sce <- sce[,sce$Celltypes_global %in% c("Enterocytes_Intermediate", "Enterocytes", "Neck-Cells", "Foveolar_Differentiated", "Intestinal_Undifferentiated", "Foveolar_Intermediate")]

# Create monocle3 object
counts.data<-counts(sce)
colnames(counts.data)<-paste(colData(sce)$Barcode, colData(sce)$Sample, sep = "_")
cell.data<-colData(sce)
rownames(cell.data)<-paste(colData(sce)$Barcode, colData(sce)$Sample, sep = "_")
gene.data<-rowData(sce)
gene.data$gene_short_name<-gene.data$Symbol
cds <- new_cell_data_set(counts.data,
                         cell_metadata = cell.data,
                         gene_metadata = gene.data)

#Get PC
cds <- preprocess_cds(cds, num_dim = 50)

# Batch correct
cds <- align_cds(cds, alignment_group = "Sample", residual_model_formula_str = "~ MT.prop" )

# Get umap
cds <- reduce_dimension(cds)

# plot using building method
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "Celltypes_global")
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "Tissue")

# Visualise Clusters
CellTypes.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                  UMAP2 = reducedDims(cds)$UMAP[,2],
                                  CellTypes = factor(cds$Celltypes_global, levels = names(cell.order)[names(cell.order) %in% unique(colData(cds)$Celltypes_global)]),
                                  Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"), "Normal", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = CellTypes, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Normal" = 1 , "Other" = 0.00)) +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=2))) +
  scale_color_manual(values = cell.order[names(cell.order) %in% unique(colData(cds)$Celltypes_global)])
# +
#   geom_text(data= aggregate(reducedDims(cds)$UMAP, list(cds$Fibroblast_clusters), mean),
#             mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(CellTypes.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_CellTypes.tiff", CellTypes.plot, device = "tiff", units="in", width=8, height=6)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_CellTypes.pdf", CellTypes.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

# Visualise Clusters
Tissue.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                               UMAP2 = reducedDims(cds)$UMAP[,2],
                               Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue), size = 0.5) + 
  theme_void() +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])

# +
#   geom_text(data= aggregate(reducedDims(cds)$UMAP, list(cds$Fibroblast_clusters), mean),
#             mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(Tissue.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue.tiff", Tissue.plot, device = "tiff", units="in", width=8, height=6)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue.pdf", Tissue.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

Tissue.plot.normal<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]),
                                      Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"), "Normal", "Other")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"),1,0.01)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("Normal" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])


print(Tissue.plot.normal)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue_normal.tiff", Tissue.plot.normal, device = "tiff", units="in", width=8, height=6)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue_normal.pdf", Tissue.plot.normal, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

Tissue.plot.IM<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                      UMAP2 = reducedDims(cds)$UMAP[,2],
                                      Tissue = factor(cds$Tissue, levels = names(colour.palet)[names(colour.palet) %in% unique(colData(cds)$Tissue)]),
                                      Alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"), "Other", "IM")
)) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black", alpha = ifelse(cds$Tissue %in% c("NSCJ", "NG", "NGB", "ND", "Ileum", "Colon", "Rectum"),0.01,1)) +
  geom_point(aes(UMAP1, UMAP2, colour = Tissue, alpha = Alpha), size = 0.5) + 
  theme_void() +
  scale_alpha_manual(values = c("IM" = 1 , "Other" = 0.00)) +
  
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=1, override.aes = list(size=3))) +
  scale_color_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)])


print(Tissue.plot.IM)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue_IM.tiff", Tissue.plot.IM, device = "tiff", units="in", width=8, height=6)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_Tissue_IM.pdf", Tissue.plot.IM, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)


cds <- cluster_cells(cds)
plot_cells(cds, color_cells_by = "partition", group_label_size = 8)

# Visualise Partitions 
partitions.plot<-ggplot(data.frame(UMAP1 = reducedDims(cds)$UMAP[,1],
                                   UMAP2 = reducedDims(cds)$UMAP[,2],
                                   Partition = factor(partitions(cds)))) +
  geom_point(aes(UMAP1, UMAP2), size = 1, color = "black") +
  geom_point(aes(UMAP1, UMAP2, colour = Partition), size = 0.5) + 
  theme_void() +
  # ylim(umap.min.max) +
  # xlim(umap.min.max) +
  guides(color=guide_legend(ncol=2, override.aes = list(size=4))) +
  geom_text(data= aggregate(reducedDims(cds)$UMAP, list(partitions(cds)), mean),
            mapping = aes(V1, V2, label = Group.1), colour = "black", size = 6)
# scale_color_manual(values = c(brewer.pal(9, "Set1"), brewer.pal(9, "Set3"))))
# scale_color_manual(values = colour.palet)

print(partitions.plot)




cds <- learn_graph(cds)

plot_cells(cds,
           color_cells_by = "Celltypes_global",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)

cds <- order_cells(cds)


pseudotime.plot <- plot_cells(cds,
                              color_cells_by = "pseudotime",
                              label_cell_groups=FALSE,
                              label_leaves=FALSE,
                              label_branch_points=FALSE,
                              graph_label_size=5,
                              cell_size = .75) + theme_void()
plot(pseudotime.plot)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_pseudotime.tiff", pseudotime.plot, device = "tiff", units="in", width=8, height=6)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Monocle_umap_pseudotime.pdf", pseudotime.plot, device = "pdf", units="in", width=8, height=6, useDingbats = FALSE)

colour.palet <- c("NE" = "#990F0F", 
                  "NSCJ" = "#CC5151",
                  "SMG" = "#FFB2B2",
                  "NG" = "#6B990F",
                  "NGB" = "#A3CC51",
                  "GM" = "#E5B17E",
                  "BE" = "#99540F",
                  "BSCJ" = "#CC8E51",
                  "NAG" = "#0F6B99",
                  "CAG" = "#2C85B2",
                  "GIM" = "#51A3CC",
                  "CIM" = "#7EC3E5",
                  "ND" = "#260F99",
                  "Ileum" = "#422CB2",
                  "Colon" = "#6551CC",
                  "Rectum" = "#8F7EE5")


Violin.plot.all <- ggplot(data = data.frame(Tissue = factor(colData(cds)$Tissue, levels = (names(colour.palet))),
                                            Tissue_type = factor(colData(cds)$Patient_status, levels = c("Healthy", "BE", "IM")),
                                            Pseudotime = pseudotime(cds)), 
                          aes(x = Tissue, y = Pseudotime)) +
  theme_cowplot(12) +
  # ylim(0,1) +
  geom_violin(aes(fill = Tissue), scale = "area", trim = TRUE) + 
  scale_fill_manual(values = colour.palet[names(colour.palet) %in% unique(colData(cds)$Tissue)]) +
  # scale_fill_manual(values = colour.palet.contributor) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  facet_wrap(vars(Tissue_type), drop = TRUE,scales = "free_x", nrow = 1)# + coord_flip()
plot(Violin.plot.all)

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Violin_pseudotime.pdf", Violin.plot.all, device = "pdf", units="in", width=5, height=8, useDingbats = FALSE)




```

## Final Figure

```{r}
lay <- rbind(c(1,2),
             c(3,4),
             c(5,5))
# c(4,4),
# c(4,4))
multiplot<-gridExtra::grid.arrange(grobs = list(Tissue.plot.normal+ theme(legend.position = "none"),
                                                Tissue.plot.IM+ theme(legend.position = "none"), 
                                                CellTypes.plot+ theme(legend.position = "none"), 
                                                pseudotime.plot, 
                                                Violin.plot.all+ theme(legend.position = "none"))
                                   , layout_matrix = lay)
ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Multiplot_Pseudotime.pdf", multiplot, device = "pdf", units="cm", width=21, height=29.7, useDingbats = FALSE)  #210 x 297

ggsave("~/Dropbox/Postdoc/2021-05-19_Gastric_IM/Results/Figures/Figure_3/Monocle/Multiplot_Pseudotime.tiff", multiplot, device = "tiff", units="cm", width=21, height=29.7)  #210 x 297

```




# End Matter

To finish get session info:


```{r Endnote, include=TRUE, echo=TRUE, warning=FALSE, eval=TRUE, message=FALSE, tidy=TRUE}
sessionInfo()
```

